---
title: "Enrichment"
author: "<h3>Author</h3>Brian M. Schilder, Bioinformatician II"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document: 
    theme: spacelab
    highlight: zenburn 
    code_folding: show 
    toc: true 
    toc_float: true
    smooth_scroll: true
    number_sections: false 
    self_contained: false
params:
  subsetGenes: "protein_coding" #FALSE
  subsetCells: 500 #FALSE
  resolution: 0.6
  resultsPath: "./Results"
  nCores: 2
  perplexity: 30
editor_options: 
  chunk_output_type: inline
---

```{r setup, message=F, results='hide'}
# Import functions
root = "./"
source(file.path(root,"MAIN.R"))

# library(enrichR) # BiocManager::install("enrichR") # devtools::install_github("wjawaid/enrichR")
# library(fgsea) 
library(gprofiler2)
 
library(dplyr) 
library(ggplot2)

load("./Data/monocle3_CDS.RData")
# print(cds)
```

# Import Gene Lists

## DEG Tables

- **dge.dx**: DGE results of dx/disease status (PD vs Controls) across all clusters.  
- **dge.clust**: DGE results of cluster (Cluster1/Canonical Monocytes vs. Cluster2/Intermediate Monocytes) across all indivdiuals (regardless of disease status).  

```{r Import: DEGs}
# ----- Across Clusters
dge.dx <- read.csv("Results/across_PD.vs.Ctrl.csv")  
# dge.mut <- read.csv("Results/across_PD.vs.GBA.csv") #No DEGs to check!
dge.clust <- read.csv("Results/across_Clust1.vs.Clust2.csv")

# ----- Within Clusters
dge.within.clust1.dx <- read.csv("./Results/within.Clust1_PD.vs.Ctrl.csv")
dge.within.clust2.dx <- read.csv("./Results/within.Clust2_PD.vs.Ctrl.csv")
dge.within.clust1.mut <- read.csv("./Results/within.Clust1_PD.vs.GBA.csv")
dge.within.clust2.mut <- read.csv("./Results/within.Clust2_PD.vs.GBA.csv") 
```

# AD and PD-associated Genes

```{r Import - AD/PD}
curatedGenes <- readxl::read_excel("Data/curated_AD-PDgene_lists.xlsx")
AD_related_genes <- readxl::read_excel(file.path(root, "Data/AD-related_genes.xlsx"))
ADgenes <- rbind(AD_related_genes, data.frame(Gene=curatedGenes$AD_panel, Category="AD") ) %>% unique()
ADgenes <- ADgenes[!is.na(ADgenes$Gene),]
# Combine PD gene lists
PDgenes <- data.frame(Gene=c("LRRK2","GBA",
                             curatedGenes$gwas_Nearest_gene_Nalls_2019, 
                             curatedGenes$QTL_Nominated_genes_Nalls_2019,
                             curatedGenes$TWAS_Garrett
                             ), Category="PD") %>% unique()
PDgenes <- PDgenes[!is.na(PDgenes$Gene),]
createDT(rbind(ADgenes, PDgenes))
```



# Curated Gene Lists

## Alzheimer's Disease

### Disease DGEs

```{r Overlap - AD vs. dx DGEs}
AD_overlap <- report_overlap(genomeSize=dim(cds)[1], 
                            list1=subset(dge.dx, q_value<=0.05)$gene_short_name, 
                            list2=ADgenes$Gene)
overlap_expression_plot(DEG_table = dge.dx, 
                        geneList = AD_overlap$overlappingGenes, 
                        title="Overlapping AD-associated Genes w/ dx DGEs")
```

### Cluster DGEs

```{r Overlap - AD vs. cluster DGEs}
AD_overlap <- report_overlap(genomeSize=dim(cds)[1], 
                            list1=subset(dge.clust, q_value<=0.05)$gene_short_name, 
                            list2=ADgenes$Gene)
overlap_expression_plot(dge.clust, 
                        geneList = AD_overlap$overlappingGenes, 
                        title="Overlapping AD-associated Genes w/ Cluster DGEs")
```




## Parkinson's Disease

### Disease DGEs

```{r Overlap - PD vs. dx DGEs}
PD_overlap <- report_overlap(genomeSize=dim(cds)[1], 
                            list1=subset(dge.dx, q_value<=0.05)$gene_short_name, 
                            list2=PDgenes$Gene)
overlap_expression_plot(dge.dx, 
                        geneList = PD_overlap$overlappingGenes, 
                        title="Overlapping PD-associated Genes w/ dx DGEs")
```

### Cluster DGEs

```{r Overlap - PD vs. cluster DGEs}
PD_overlap <- report_overlap(genomeSize=dim(cds)[1], 
                            list1=subset(dge.clust, q_value<=0.05)$gene_short_name, 
                            list2=PDgenes$Gene)
overlap_expression_plot(dge.clust, 
                        geneList = PD_overlap$overlappingGenes, 
                        title="Overlapping PD-associated Genes / Cluster DGEs")
```

  
 

# GO Enrichment  

- [gprofiler2](https://biit.cs.ut.ee/gprofiler/page/r)  
- [Enrichr](https://amp.pharm.mssm.edu/Enrichr/)
    + enrichR R package not currently working (8.14.2019). Providing the [link to the online results here](https://amp.pharm.mssm.edu/Enrichr/enrich?dataset=d634d1a5b290e873498f65676c894b33) for now. 

## Across Clusters  

### dx DGEs

```{r Across Clusters - dx}
query <- subset(dge.dx, q_value<0.05)$gene_short_name %>% as.character()
gostres <- gprofiler2::gost(query = query, organism = "hsapiens")
gprofiler2::gostplot(gostres)
createDT(gostres$result)
```

### mut DGEs

```{r Across Clusters - mut}
print("No DEGs were identified from this comparison!")
```

### Cluster DGE

- DGE from Cluster 1 (canonical) vs. Cluster 2 (intermediate) comparison.

```{r Across Clusters - cluster}
query <- subset(dge.clust, q_value<0.05)$gene_short_name %>% as.character()
gostres <- gprofiler2::gost(query = query, organism = "hsapiens")
gprofiler2::gostplot(gostres)
createDT(gostres$result)
```

## Within Clusters

### dx DGEs

#### Cluster 1

```{r Within Cluster1 - dx}
query <- subset(dge.within.clust1.dx, q_value<0.05)$gene_short_name %>% as.character()
gostres <- gprofiler2::gost(query = query, organism = "hsapiens")
gprofiler2::gostplot(gostres)
createDT(gostres$result)
```

#### Cluster 2

```{r Within Cluster2 - dx}
query <- subset(dge.within.clust2.dx, q_value<0.05)$gene_short_name %>% as.character()
gostres <- gprofiler2::gost(query = query, organism = "hsapiens")
gprofiler2::gostplot(gostres)
createDT(gostres$result) 
```

### mut DGEs

#### Cluster 1

```{r Within Cluster1 - mut}
query <- subset(dge.within.clust1.mut, q_value<0.05)$gene_short_name %>% as.character()
gostres <- gprofiler2::gost(query = query, organism = "hsapiens")
gprofiler2::gostplot(gostres)
createDT(gostres$result)  
```

#### Cluster 2

```{r Within Cluster2 - mut}
query <- subset(dge.within.clust2.mut, q_value<0.05)$gene_short_name %>% as.character()
gostres <- gprofiler2::gost(query = query, organism = "hsapiens")
gprofiler2::gostplot(gostres)
createDT(gostres$result)
```



