---
title: "PD_scRNAseq: Monocle3"
subtitle: "Preprocessing + Dimensionality Reduction + Pseudotime + Clustering + Cell Type Identification + Gene Modules"
author: "<h3>Author</h3>Brian M. Schilder, Bioinformatician II"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document: 
    theme: spacelab
    highlight: zenburn 
    code_folding: show 
    toc: true 
    toc_float: true
    smooth_scroll: true
    number_sections: false 
    self_contained: true
params:
  subsetGenes: "protein_coding" #FALSE
  subsetCells: 500 #FALSE
  resolution: 0.6
  resultsPath: "./Results"
  nCores: 2
  perplexity: 30
editor_options: 
  chunk_output_type: inline
---

# Setup 

```{r setup, dpi = 600, warning=F, message=F}
start_time <- Sys.time() 

#### #### Load Objects & Functions #### #### 
######################################################
# Import functions
root = "./"
source(file.path(root,"MAIN.R"))
import_parameters(params)


######################################################
#### #### PACKAGES #### #### 
######################################################

print("Written using: Seurat version* 2.3.4 2018-07-17")
# https://satijalab.org/seurat/install.html
# source("http://bit.ly/archived-seurat")
library(Seurat) 
paste("Seurat", packageVersion("Seurat")) 

# library(monocle) # BiocManager::install("monocle")   
# paste("monocle", packageVersion("monocle")) 
## WARNING! There's currently bugs in the Bioconductor installation of Monocle3, so install the dev version from GitHub instead (see https://github.com/cole-trapnell-lab/monocle3/issues/272)
library(monocle3); # devtools::install_github('cole-trapnell-lab/monocle3', ref="develop")
paste("monocle3", packageVersion("monocle3")) 

# library(garnett) # devtools::install_github("cole-trapnell-lab/garnett")
# paste("garnett", packageVersion("garnett"))
 
library(cowplot) 
library(ggplot2)
library(dplyr) 
library(data.table) 
library(readxl) 
library(reshape2)
library(ggrepel)
library(plotly)
######################################################


# Exporting 3D plots 
knitr::knit_hooks$set(webgl = rgl::hook_webgl)
dge_limit <- F # 100
nCores <- 4#parallel::detectCores()
set.seed(2019)
```


# Preprocessed Data

+ A cds object with 14k+ protein-coding genes and 19k+ cells (after filtering with Seurat::FilterCells()).  
+ Normalization, PCA, UMAP, clustering and pseudotime have already been done on this object.

```{r Preprocessed Data}
# load("Data/seurat_object_add_HTO_ids.Rdata")
load(file.path("Data", "monocle3_CDS.RData"))

# write.csv(pData(cds), file="Results/Monocle3/processed_metadata.csv")
# R.utils::gzip("Results/Monocle3/processed_metadata.csv")
```

# Preprocessing

## Seurat/2.3.4 to Monocle3
- Below is the pre-processed data, and the pipeline to get it to the same stage as "monocle3_CDS.RData".  
- Only include protein-coding genes. Otherwise, the clusters don't map as well onto the UMAP plots.  
- Only need to run if not using preprocessed ``cds`` object.  

```{r Seurat/2.3.4 to Monocle3, eval=F}
load(file.path(resultsPath, "3-11-2019/scRNAseq_results.RData"))
# load("./Data/seurat_object_add_HTO_ids.Rdata")
DAT <- Seurat::UpdateSeuratObject(DAT)
# remove(seurat.obj)

# DAT <- readRDS("./Data/processed_seurat_obj.RDS")
# Find variable genes
# DAT <- FindVariableFeatures(object = DAT, 
#                             mean.function = ExpMean, 
#                             dispersion.function = LogVMR,
#                             x.low.cutoff = 0.0125, 
#                             x.high.cutoff = 3, 
#                             y.cutoff = 0.5)
var.genes <- Seurat::VariableFeatures(object = DAT)
var.genes <- unique(c(var.genes, "CD14", "FCGR3A"))

biotypes <- get_biotypes(DAT)
proteins <- biotypes[biotypes$gene_biotype=="protein_coding",]$hgnc_symbol %>% droplevels()
# protDAT <-  subset_seurat(DAT, genes.use = proteins) 
 
cds <- seurat_to_monocle(seurat_object = DAT, 
                         seurat_version = 3,
                         monocle_version = 3)
cds <- cds[levels(proteins),]
var.genes <- var.genes[var.genes %in% row.names(cds)]
```
 
## Normalization & PCA

### Curated genes lists

Import curated gene lists from Elisa Navarro.
Feed these into the `use_genes` parameter of the `preprocess_cds()` function.
This will be necessary to be able to create gene modules with them later on.

```{r Curated Gene Lists}
lists_path <-"./Results/curated_modules/"
gene.lists <- lapply(list.files(lists_path, full.names = T, pattern = "*.csv$"), function(x){
  group.name <- gsub(".csv$","",basename(x))
  dat <- data.table::fread(x)
  dat <- dat %>% dplyr::mutate(module = group.name)
  colnames(dat)[1] <- "gene.symbol"
  return(data.table::data.table(dat))
}) %>% data.table::rbindlist()


# Separate MT-XXX genes from mito
## (separate from other modules too?)
mtx <- readxl::read_excel("./Results/curated_modules//MT-XXX_genes.xlsx")
gene.lists[gene.lists$module=="mito" & (gene.lists$gene.symbol %in% mtx$gene.symbol),"module"] <- "MT-XXX"  
if(sum(!(mtx$gene.symbol %in% gene.lists$gene.symbol)) >0){
  print("All MT-XXX already in dataframe.")
}
gene.lists <- unique(subset(gene.lists, gene.symbol %in% row.names(fData(cds))))
use_genes <- unique(gene.lists$gene.symbol)

data.table::fwrite(gene.lists, "./Results/curated_modules//curatedModules_genelists.tsv",sep="\t")
```


```{r Normalize Data}
cds <- monocle3::preprocess_cds(cds, 
                                # num_dim = 30, #100 by default
                                return.only.var.genes=F,
                                use_genes = var.genes,
                                residual_model_formula_str = "~ nUMI + percent.mito")
pc_var <- monocle3::plot_pc_variance_explained(cds)
print(pc_var)
ggsave("Results/Monocle3/PCA_variance_explained.pdf",plot = pc_var, dpi = 400, height=5)
```


## Dimensionality Reduction

Using UMAP.

```{r Dimensionality Reduction}
# c("UMAP", "tSNE", "PCA", "LSI")
cds <- monocle3::reduce_dimension(cds,
                                  preprocess_method = "PCA", 
                                  reduction_method = "UMAP", 
                                  max_components = 3, 
                                  # umap.fast_sgd = T,
                                  # cores=nCores
                                  ) 
gg_markers <- monocle3::plot_cells(cds, genes = c("CD14","FCGR3A"),
                                   show_trajectory_graph = F, alpha = .8, cell_size = .75)
print(gg_markers)
ggsave("Results/Monocle3/UMAP_CD14_FCGR3A.pdf", plot = gg_markers, dpi = 400, width = 12)
```


# Clustering

- Unsupervised clustering of cells is a common step in many single-cell expression workflows. In an experiment containing a mixture of cell types, each cluster might correspond to a different cell type. - This function takes a cell_data_set as input, clusters the cells using Louvain community detection, and returns a cell_data_set with internally stored cluster assignments. 
- In addition to clusters this function calculates partitions, which represent superclusters of the Louvain communities that are found using a kNN pruning method. Cluster assignments can be accessed using the clusters function and partition assignments can be accessed using the partitions function.

- Using only the topN variable genes to cluster

## Find Variable Genes

```{r Find Variable Genes} 
# Method 1
# variable.genes <- cds@preprocess_aux@listData$gene_loadings[,1:3] %>%
#   abs() %>% rowSums() %>% base::sort(decreasing = T)
# head(variable.genes) 
# Seurat v2 function
# varDAT <- Seurat::FindVariableGenes(object = protDAT, 
#                              mean.function = ExpMean, 
#                              dispersion.function = LogVMR, 
#                              selection.method ="dispersion", do.plot = T, 
#                              top.genes = 2000)
# var.genes <- varDAT@var.genes


# Seurat v3 function
# protNEW <- UpdateSeuratObject(protDAT)  
# varDAT <- Seurat:::FindVariableFeatures(object = protNEW,
#                                         selection.method = "vst",
#                                         mean.function = "FastExpMean", 
#                                         dispersion.function = "FastLogVMR",
#                                         nfeatures = 2000)
# var.genes <- varDAT@assays$RNA@var.features
# head(var.genes)
# 

# Using only monocle3 
length(unique(row.names(cds@preprocess_aux$gene_loadings[,1:10])))
var.genes <- names(sort(rowSums(abs(cds@preprocess_aux$gene_loadings[,1:10]), na.rm = T), T))#[1:2000] 
```

## Louvain Clustering

*NOTE*: In the new version of **cluster_cells**, you can now leave the **resolution** parameter blank and it will automatically optimize the resolution. The default used to be **1e-4**.
```{r Louvain Clustering, webgl=TRUE} 
cds <- monocle3::cluster_cells(cds,
                               resolution = 2e-5,#c(10^seq(-6,-1)), # Do NOT set to high number
                               reduction_method = "UMAP", 
                               cores = nCores,
                               # clustering_genes = var.genes
                               )
# Add cluster info to metadata (for easier DGE analysis)
pData(cds)$Cluster <- monocle3::clusters(cds)
pData(cds)$Partition <- monocle3::partitions(cds)

p2d <- plot_cells(cds,
                  # x = 1, y = 1,  
                  color_cells_by = "cluster", 
                  show_trajectory_graph = F, 
                  group_label_size = 5, 
                  cell_size = .75, 
                  alpha = .5, 
                  rasterize = T)
p2d
ggsave(plot = p2d, filename = "./Results/Monocle3/UMAP_clusters.pdf", dpi = 600)

```

### 3D UMAP Clusters

```{r 3D Plot}
# 3D plot
p3d <- plot_cells_3d(cds, color_cells_by = "cluster", show_trajectory_graph = F)
p3d

# for(i in seq(0,6.3,by=0.1)){
#   cam.zoom = 2
#   ver.angle = 0
#   graph <- p3d %>% plotly::layout(
#   camera = list(eye = list(x = cos(i)*cam.zoom,y = sin(i)*cam.zoom, z=0.2),
#                                             center = list(x = 0,
#                                                           y = 0,
#                                                           z = 0
#                                                           )
#                                             )
#   )
# cat("Now rendering iteration:", i,"\n")
# plotly_IMAGE(graph,
#          width = 1200,
#          height = 1050,
#          format = "png",
#          username="bschilder",
#          key="vCq0X23Hws3tIHhJpU9I",
#          scale = 1,
#          out_file = paste("./plotly_test","png", sep="."))
# }
# Save each cluster as a separate OBJ file that can later be 
# imported into a 3D modeling platform (e.g. Blender, Meshlab) for more advanced visualization
## save_3D_clusters(cds)
```

# Pseudotime

```{r Pseudotime} 
# Learn
cds <- monocle3::learn_graph(cds)
# Order cells
root_pr_nodes = get_earliest_principal_node(cds, variable = "dx", variable_value = "PD")
cds <- monocle3::order_cells(cds, root_pr_nodes = root_pr_nodes) 

monocle3::plot_cells(cds)
monocle3::plot_genes_in_pseudotime(cds[c("CD14","FCGR3A"), ] )
```

### UMAP Plots {.tabset .tabset-fade .tabset-pills}

Disease status (dx), mutation status (mut), and individual ID (ID) show good mixture across clusters.
```{r}
p <- monocle3::plot_cells(cds, label_cell_groups=F, show_trajectory_graph = F)
p

p_dx <- monocle3::plot_cells(cds,color_cells_by = "dx", label_cell_groups = F,
                             show_trajectory_graph = F, cell_size = .75, alpha = .8)
print(p_dx)
ggsave(plot = p_dx, filename = "./Results/Monocle3/UMAP_dx.pdf", dpi = 600)


```



```{r UMAP Plots, results="asis"}
for(v in c("dx","mut","ID")){
  cat("\n")
  cat("####",v,"\n")
  p <- monocle3::plot_cells(cds, color_cells_by=v, label_cell_groups=F, show_trajectory_graph = F)
  print(p)
  cat("\n")
} 
```



# Cell-type Identication

## Garnett

Use a pre-trained classifier from Pline et al.

```{r Garnett}
library(org.Hs.eg.db)
columns(org.Hs.eg.db)

# load(url("https://cole-trapnell-lab.github.io/garnett/classifiers/hsPBMC")) 
load("./Data/Garnett/hsPBMC")
cds = garnett::classify_cells(cds, 
                               hsPBMC,
                               db = org.Hs.eg.db,
                               cluster_extend = T,
                               cds_gene_id_type = "SYMBOL")
monocle3::plot_cells(cds,
                     group_cells_by="cluster",
                     color_cells_by="cluster_ext_type",  # cell_type
                     show_trajectory_graph = F )
```


## Gene Expression Plots

### Violin

```{r QC violin}
pdata <- data.frame(pData(cds))
pdata$dx <- factor(pdata$dx, levels=c("PD","control"), ordered = T)
p_viol_qc <- ggplot(pdata, aes(x=nGene, fill=dx)) + 
  geom_histogram(alpha=.75, position = 'identity') +
  theme_bw() +
  labs(y="Frequency")
print(p_viol_qc)
ggsave("Results/Monocle3/histogram_dx.pdf", p_viol_qc, dpi=400, height=5)


mito_dx_clust  <- ggplot(pdata, aes(x=dx, y=percent.mito, fill=Cluster)) + 
  geom_boxplot() +
  theme_bw()
mito_dx_clust
ggsave("Results/Monocle3/percentMito_dx_cluster_boxplot.pdf", mito_dx_clust, 
       dpi=400, height=5)

```


```{r Gene Expression Plots - Violin}
# Violin
p_viol_markers <- monocle3::plot_genes_violin(cds[c("CD14","FCGR3A"),], group_cells_by = "dx")
```

## UMAP

```{r Gene Expression Plots - UMAP}
# Key Markers
monocle3::plot_cells(cds, 
                     genes=c("CD14","FCGR3A"), 
                     group_cells_by = "cluster",
                     show_trajectory_graph = F) 
```



# Cell Type Proportions 

- Check the proportion of of cells within each disease group that belong to cluster. This will help determine whether any PD vs. Control DEGs are reflective of cell proportion differences.

## dx

```{r Cluster Proportions - dx}
meta <- data.frame(pData(cds)) 
# By dx
cell_proportions <- meta %>%  
                   group_by(dx, Cluster) %>% 
                   summarise(Proportion=n()) %>% 
                   group_by(dx) %>% 
                   mutate(Proportion=Proportion/sum(Proportion)) 
createDT(cell_proportions)
prop.dx <- ggplot(cell_proportions, aes(x=dx, y=Proportion, fill = Cluster)) + 
              geom_col() + # position="dodge"
              ylab("Proportion of Cells") + 
  theme_bw()
print(prop.dx)
ggsave("Results/Monocle3/cell_proportions.pdf",plot = prop.dx, dpi = 400, height=6, width=5)



clust.key <- list("1"="Canonical Monocytes", "2"="Intermediate Monocytes")
for(clust in 1:2){ 
clust1.diff <- (subset(cell_proportions, Cluster==clust & dx=="PD")$Proportion - subset(cell_proportions, Cluster==clust & dx=="control")$Proportion)*100
# print(paste("There is a",round(clust1.diff, 2),"% difference in the number of Cluster",clust,"cells (",clust.key[[as.character(clust)]], ") in Controls compared to PD."))
} 

subset(meta, dx=="control")
avg.cells.per.ID <- meta %>% dplyr::group_by(ID, dx, Cluster) %>% 
  tally() %>% 
  dplyr::group_by(dx, Cluster) %>% 
  summarise(N = mean(n))
avg.cell.filt <- avg.cells.per.ID[avg.cells.per.ID$Cluster %in% c(1,2),] %>% droplevels()


avg.table <- xtabs(N ~ dx+Cluster, data=avg.cell.filt)
chisq.test(avg.table)
```

## mut

```{r Cluster Proportions - mut} 
# By mut
cell_proportions <- meta %>%  
                   group_by(mut, Cluster) %>% 
                   summarise(Proportion=n()) %>% 
                   group_by(mut) %>% 
                   mutate(Proportion=Proportion/sum(Proportion)) 
createDT(cell_proportions)
prop.mut <- ggplot(cell_proportions, aes(x=mut, y=Proportion, fill = Cluster)) + 
              geom_col() + # position="dodge"
              ylab("Proportion of Cells") 
print(prop.mut) 
```

## individual

```{r Cluster Proportions - individual} 
# By mut
cell_proportions <- meta %>%  
                   group_by(dx, Cluster, ID) %>% 
                   summarise(Proportion=n()) %>% 
                   group_by(ID) %>% 
                   mutate(Proportion=Proportion/sum(Proportion)) 
createDT(cell_proportions)
prop.ID <- ggplot(cell_proportions, aes(x=ID, y=Proportion, fill = Cluster)) + 
              geom_col() + # position="dodge"
              ylab("Proportion of Cells") + 
              facet_grid("~dx") + 
              theme(axis.text.x = element_text(angle = 45))
print(prop.ID) 
```





# Save Checkpoint

Save R object and run memory-intensive DGE analyses on computing cluster.

```{r Save Checkpoint} 
# saveRDS(cds, file="./Data/monocle3_CDS.RData")
save(cds, var.genes, file = "./Data/monocle3_CDS.RData")
```

