---
title: "Monocyte Subtypes" 
author: "<h3>Author</h3>Brian M. Schilder, Bioinformatician II"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document: 
    theme: spacelab
    highlight: zenburn 
    code_folding: show 
    toc: true 
    toc_float: true
    smooth_scroll: true
    number_sections: false 
    self_contained: true
params:
  subsetGenes: "protein_coding" #FALSE
  subsetCells: 500 #FALSE
  resolution: 0.6
  resultsPath: "./Results"
  nCores: 2
  perplexity: 30
---

# Setup 

```{r setup, dpi = 600}
# Import functions
root = "./"
source(file.path(root,"general_functions.R"))
import_parameters(params)
# load(file.path(resultsPath, "scRNAseq_results.RData"))
resultsPath <- "Results/subsetGenes-protein_coding__subsetCells-F__Resolution-0.2__perplexity-40__nCores-4"
load(file.path(resultsPath,"scRNAseq_results.RData")) 
```

## Load Libraries

```{r Load Libraries, message=F, warning=F}
library(Seurat)
library(cowplot)
library(ggplot2)
library(dplyr) 
library(data.table) 
library(readxl) 
library(reshape2)
library(ggrepel)
library(plotly)

library(GeneOverlap) # BiocManager::install("GeneOverlap") 
library(enrichR) #BiocManager::install("enrichR")

## Upgrade to alpha (development) version of Monocle, as this version has been optimized for large datasets
if("DDRTree" %in% rownames(installed.packages())==F){
  devtools::install_github("cole-trapnell-lab/DDRTree", ref="simple-ppt-like")
}
if("graph" %in% rownames(installed.packages())==F){
 devtools::install_github("cole-trapnell-lab/L1-graph") 
}
if("reticulate" %in% rownames(installed.packages())==F){
 install.packages("reticulate")
}

library(reticulate)
if(py_module_available("umap-learn")==F){
  reticulate::py_install('umap-learn')# , pip = T, pip_ignore_installed = T # Ensure the latest version of UMAP is installed 
}
if(py_module_available("louvain")==F){
  reticulate::py_install("louvain")
  system("pip install louvain --user")
}
if("monocle" %in% rownames(installed.packages())==F){
  devtools::install_github("cole-trapnell-lab/monocle-release", ref="monocle3_alpha") 
}
if("ggrastr" %in% rownames(installed.packages())==F){
  devtools::install_github("VPetukhov/ggrastr")
}
library(monocle) # BiocManager::install("monocle") 
# BiocManager::install(c('DelayedArray', 'DelayedMatrixStats', 'org.Hs.eg.db', 'org.Mm.eg.db'))
library(org.Hs.eg.db)
library(garnett) # devtools::install_github("cole-trapnell-lab/garnett")
 
#  
# if(getwd()=="/Users/schilder/Desktop/PD_scRNAseq"){
#   allGenes <- F
# }else{allGenes <- T}
# allGenes
```

# Monocle 

## Convert and Normalize

* Convert from Seurat object to CDS object and re-normalize the data from scratch (importCDS currently only converts the raw.data, not the scale.data).
* Alternatively, you can do your normalization in Seurat and then construct a new CDS object manually with the scale.data from the Seurat object.
```{r Convert and Normalize}
# Pass TRUE if you want to see progress output on some of Monocle 3's operations
DelayedArray:::set_verbose_block_processing(TRUE)
# Passing a higher value will make some computations faster but use more memory. Adjust with caution!
options(DelayedArray.block.size=1000e6)

# Construct CDS object manually
# mDAT <- newCellDataSet(cellData = DAT@scale.data, 
#                        featureData = AnnotatedDataFrame( data.frame(gene_short_name=row.names(DAT@scale.data), 
#                                                                     row.names = row.names(DAT@scale.data))
#                          ),
#                        phenoData = AnnotatedDataFrame(DAT@meta.data))
# mDAT <- estimateSizeFactors(mDAT) #DESeq2?
# mDAT <- preprocessCDS(mDAT, method = "PCA", num_dim = 10, norm_method = "none")
# 

## Construct CDS object automaticaly
### NOTE!: importCDS takes only the raw.data (not scale.data)
mDAT <- monocle::importCDS(DAT,  import_all = T)
mDAT <- estimateSizeFactors(mDAT) #DESeq2?
# mDAT <- estimateDispersions(mDAT)
mDAT <- preprocessCDS(mDAT, num_dim = 10, residualModelFormulaStr = "~ nUMI + percent.mito + ID")
```


## Dimensionality Reduction

```{r Dimensionality Reduction}
set.seed(1) # Monocle sets their seed to 2016 by default (Seurat sets it to 1)
# Options: c("UMAP", "tSNE", "DDRTree", "ICA", "none")
# mDAT <- reduceDimension(mDAT, reduction_method = 'UMAP', n_threads =  parallel::detectCores())
# monocle:::UMAP( )
mDAT <- reduceDimension(mDAT, reduction_method = 'tSNE', 
                        perplexity = 30,
                        scaling = F,
                        num_threads = parallel::detectCores()) 

col_vector_origin <- c("#db83da",
              "#53c35d",
              "#a546bb",
              "#83b837",
              "#a469e6",
              "#babb3d",
              "#4f66dc",
              "#e68821",
              "#718fe8",
              "#d6ac3e",
              "#7957b4",
              "#468e36",
              "#d347ae",
              "#5dbf8c",
              "#e53e76",
              "#42c9b8",
              "#dd454a",
              "#3bbac6",
              "#d5542c",
              "#59aadc",
              "#cf8b36",
              "#4a61b0",
              "#8b8927",
              "#a24e99",
              "#9cb36a",
              "#ca3e87",
              "#36815b",
              "#b23c4e",
              "#5c702c",
              "#b79add",
              "#a55620",
              "#5076af",
              "#e38f67",
              "#85609c",
              "#caa569",
              "#9b466c",
              "#88692c",
              "#dd81a9",
              "#a35545",
              "#e08083",
              "#17becf",
              "#9edae5")
```


## Clustering

* "Here __res__ represents the resolution parameter (range from 0-1) for the louvain clustering. Values between 0 and 1e-2 are good, __bigger values give you more clusters__. Default is set to be 1e-4. louvain_iter represents the number of iterations used for Louvain clustering. The clustering result gives the largest modularity score that will be used as the final clustering result. The default is 1."
```{r Clustering}  
plotDR <- function(resDAT, metavar="Cluster", title=""){ 
  col_vector <- col_vector_origin[1:length(unique(as.character(pData(mDAT)$dx)))]
  names(col_vector) <- unique(as.character(pData(mDAT)$dx))
  p <- plot_cell_clusters(resDAT,
                     color_by = metavar,
                     cell_size = 0.75,
                     show_group_id = T)  +
    theme(legend.text=element_text(size=6)) +
    theme(legend.position="right")+ 
    labs(title = paste(title))
  print(p)
}

test_resolutions <- function(mDAT, resolutions=seq(0, 1e-4, length.out = 10)){
  for(res in resolutions){
    try({
      res_title <- paste("Resolution =",res)
      cat("\n",res_title,"\n")
     resDAT <- clusterCells(mDAT, res=res,
                        method = "louvain", 
                        verbose = F, 
                        cores = parallel::detectCores())
    plotDR(resDAT, title=res_title)  
    }) 
  } 
}


test_resolutions(mDAT)
# Select final resolution
selected_res <- 4e-05
mDAT <- clusterCells(mDAT, res=selected_res,
                        method = "louvain", 
                        verbose = T, cores = parallel::detectCores()) 

plotDR(mDAT, "Cluster")
plotDR(mDAT, "dx")
plotDR(mDAT, "mut")
# CLUSTERING FROM SEURAT
plotDR(mDAT, "post_clustering")
```

## Identify Cell Types with Garnett

```{r Identify Cell Types with Garnett} 
# generate size factors for normalization later
# Get pre-trained PBMC classifer 
load(file.path(root, "Data/Garnett/hsPBMC")) # Download from: https://cole-trapnell-lab.github.io/garnett/classifiers/hsPBMC  

mDAT <- garnett::classify_cells(mDAT, 
                           classifier = hsPBMC,
                           db = org.Hs.eg.db,
                           # cluster_extend = TRUE,
                           cds_gene_id_type = "SYMBOL")
table(pData(mDAT)$cell_type)
table(pData(mDAT)$cluster_ext_type) 
# Get feature genes for each cell type
feature_genes <- garnett::get_feature_genes(classifier = hsPBMC,
                                   node = "root",
                                   db = org.Hs.eg.db,
                                   convert_ids = F)
head(feature_genes)  

plotDR(mDAT, metavar = "cell_type") +  facet_wrap(~dx)
```

## Gene Expression

### Monocyte Markers

```{r Monocyte Markers}
plot_cell_clusters(mDAT, markers = c("CD14","FCGR3A"), cell_size = 0.8)
```

### PD Genes

```{r PD Genes}
plot_cell_clusters(mDAT, markers = c("LRRK2","GBA", overlappingGenes_PD), cell_size = 0.75)
```

### AD Genes

```{r AD Genes} 
plot_cell_clusters(mDAT, markers = overlappingGenes_AD, cell_size = 0.75)
```

## Selected Genes

```{r Selected Genes}
plot_cell_clusters(mDAT, markers = c("MS4A4A","MS4A6A"), cell_size = 0.75) 
```




## Pseudotime

```{r Pseudotime}  
# subsetCDS(mDAT, )
mDAT <- partitionCells(mDAT)
mDAT <- learnGraph(mDAT,  RGE_method = 'SimplePPT')

plot_cell_trajectory(mDAT, color_by = "Cluster") 
plot_cell_trajectory(mDAT, color_by = "dx") 
plot_cell_trajectory(mDAT, color_by = "mut") 
plot_cell_trajectory(mDAT, color_by = "cell_type") 

plot_3d_cell_trajectory(mDAT,
                        color_by="cell_type",
                        webGL_filename= paste(getwd(), "/trajectory_3D.html", sep=""), 
                        show_backbone=TRUE,
                        useNULL_GLdev=TRUE)
plot_3d_cell_trajectory(mDAT, markers = c('FCGR3A'),
                        webGL_filename=paste(getwd(), "/beta_globin.html", sep=""),
                        show_backbone=TRUE,
                        useNULL_GLdev=TRUE)
```

## Differential Expression

```{r Differential Expression}
DEG_df <- differentialGeneTest(mDAT, fullModelFormulaStr = "~Cluster",
                               cores = parallel::detectCores())
```

## Convert back to Seurat

```{r Convert back to Seurat} 
DAT <- Seurat::AddMetaData(DAT, pData(mDAT)[c("garnett_cluster","cell_type","cluster_ext_type")])
write.csv(DAT@meta.data, file.path(resultsPath, "garnett_results.csv"), quote = F,row.names = F)
```






# Biomarker Expression

```{r Biomarker Expression}
markerList <- c("CD14", "FCGR3A")
markerData <- DAT@scale.data[row.names(DAT@scale.data) %in% markerList,] %>% t() %>% data.frame()  
## Efficiently merge using data.table
dt1 <- data.table(markerData, keep.rownames = "Cell", key = "Cell") %>% copy()
dt2 <- data.table(DAT@meta.data[c("cell_type","post_clustering")], keep.rownames = "Cell", key = "Cell") %>% copy()
markerDT <- dt1[dt2]
 
ggplot(data = markerDT, aes(x=CD14, y=FCGR3A) ) + 
  stat_density_2d(aes(fill = ..level..), geom = "polygon", colour="purple", bins = 100, size=.1) +
  geom_point(aes(color=as.factor(cell_type)), shape=16, stroke=0, size=2, alpha=.1) + 
  guides(colour = guide_legend(title="cell_type",override.aes = list(alpha = 1))) 
```



## tSNE Metadata Plots

```{r Plot tSNE}
tSNE_metadata_plot <- function(var, labSize=12){ 
  # Metadata plot  
  p1 <- Seurat::TSNEPlot(DAT, do.return = T,  do.label = T,  group.by = var,label.size = labSize,
                 plot.title=paste(var), vector.friendly=T) +
    theme(legend.position = "top") + 
    scale_color_brewer( palette = "Dark2",  aesthetics = aes(alpha=.5)) 
     
  # t-SNE clusters plot
  p2 <- Seurat::TSNEPlot(DAT, do.return = T, do.label = T,label.size = labSize,
                 plot.title=paste("Unsupervised Clusters"), vector.friendly=T)  +
    theme(legend.position = "top") + 
    scale_color_brewer( palette = "Set1", aesthetics = aes(alpha=.5))  
  print(cowplot::plot_grid(p1,p2))
}   
```

### Garnett cell_type

```{r Garnett cell_type}
tSNE_metadata_plot("cell_type")
```

### Garnett cluster_ext_type

```{r Garnett cluster_ext_type}
tSNE_metadata_plot("cluster_ext_type")
```

### mut

```{r mut}
tSNE_metadata_plot("mut")
```

## tSNE FeaturePlots

### Monocyte Markers

```{r tSNE FeaturePlots: Monocyte Markers}
FeaturePlot(DAT,features.plot =  c("CD14","FCGR3A"),  
            cols.use = c("grey","red","blue","purple"),
            dark.theme = T, nCol = 2, overlay = T, no.legend = F) 
```

### PD Genes

```{r tSNE FeaturePlots: PD Genes} 
FeaturePlot(DAT,features.plot = c("LRRK2", "GBA"),  
            cols.use = c("grey","purple","cyan","blue"),
            dark.theme = T, nCol = 2, overlay = T, no.legend = F) 
```

### MS4A4A & MS4A6A

```{r tSNE FeaturePlots: MS4A4A & MS4A6A}
FeaturePlot(DAT, features.plot =  c("MS4A4A","MS4A6A"), 
            cols.use = c("grey","red","green","blue"),
            dark.theme = T, nCol = 2, overlay = T, no.legend = F)
```



# AD/PD-related Gene Expression

Q: Are AD/PD-related gene expression more prevalent in classical or intermediate monocyte subtypes? 
```{r AD/PD-related Gene Expression}
# Combine AD gene lists
curatedGenes <- read_excel("Data/curated_AD-PDgene_lists.xlsx")
AD_related_genes <- read_excel(file.path(root, "Data/AD-related_genes.xlsx"))
ADgenes <- rbind(AD_related_genes, data.frame(Gene=curatedGenes$AD_panel, Category="AD") ) %>% unique()
# Combine PD gene lists
PDgenes <- data.frame(Gene=c("LRRK2","GBA",
                             curatedGenes$gwas_Nearest_gene_Nalls_2019, 
                             curatedGenes$QTL_Nominated_genes_Nalls_2019,
                             curatedGenes$TWAS_Garrett
                             ), Category="PD") %>% unique()


MonocyteSubtype.markers <- read.csv(file.path(root, resultsPath, "MonocyteSubtype.markers.csv"), row.names = 1)
MonocyteSubtype.markers_sig <- subset(MonocyteSubtype.markers, p_val_adj <= 0.05) 
createDT(MonocyteSubtype.markers, "Full list of DEGs: Cluster 0 vs. Cluster 1")
```

## Test Gene Overlap

### Functions

#### Gene Overlap Functions

* _"The GeneOverlap class formulates the problem as testing whether two variables are independent, which can be represented as a contingency table, and then uses Fisherâ€™s exact test to find the statistical significance."_
* Low p-value means the overlap is highly significant. 
* [Documentation](http://bioconductor.org/packages/release/bioc/vignettes/GeneOverlap/inst/doc/GeneOverlap.pdf)
```{r Test Gene Overlap} 
report_overlap <- function(DAT, list1, list2){ 
  genomeSize <- dim(DAT@raw.data)[1]
  go.obj <- newGeneOverlap(listA = list1, listB = list2, genome.size = genomeSize )
  go.obj <- testGeneOverlap(go.obj)
  print(go.obj) 
  
  
  overlappingGenes <- getIntersection(go.obj)
  percent_of_targetGenes <- length(overlappingGenes) / length(list2)*100
  percent_of_DEGs <- length(overlappingGenes) / length(list1)*100
  targetGenes_DEGs <- list2[overlappingGenes %in% list2]
  
  cat("\n",round(percent_of_targetGenes, 2),"% of the provided genes (",length(overlappingGenes),"/",length(list2),") are differentially expressed between Canonical vs. Intermediate monocyte subtypes.")
   
  cat("\n",round(percent_of_DEGs, 2),"% of the DEGs between Canonical vs. Intermediate  monocyte subtypes (",
      length(overlappingGenes),"/",length(list1),") are in the provided gene list.")
  
  cat("\n-------------------------------------------------------",
      "\n\n********** Enrichment p-value =",go.obj@pval,"**********\n\n")
  return(overlappingGenes) 
} 
```

#### Overlap Plot Functions

```{r Overlap Plot Functions}
clustDAT <- SubsetData(DAT, subset.name = "post_clustering", accept.value = c(0,1), do.scale = F)

# PLOT FUNCTIONS
overlap_heatmap <- function(clustDAT, geneList, title="Overlapping Genes:\nGene List vs. DGE Genes"){ 
  markerDF  <- get_markerDT(clustDAT, markerList = geneList, rawData = T)
  markerMatrix <- reshape2::acast(markerDF, Gene~post_clustering, value.var="Expression",
                                  fun.aggregate = mean, drop = F, fill = 0)
  # Heatmap.2 
  my_palette <- colorRampPalette(c("purple", "black", "cyan"))(n = 1000)
  hmap <- gplots::heatmap.2(markerMatrix, xlab = "Cluster", dendrogram = "row",
                    col = my_palette, tracecol = "gray", srtCol = 0, offsetCol=1.5, vline=T, 
                    trace='none', key.title=NA, key.ylab = "Expression", colsep=T, sepwidth = 0.01, 
                    main = title) 
}


make_fractionDF <- function (MonocyteSubtype.markers_sig, geneList, allGenes=F){
  MonocyteSubtype.markers_sig$Gene <- row.names(MonocyteSubtype.markers_sig)
 if(allGenes==F){
   geneDF <- subset(MonocyteSubtype.markers_sig, Gene%in%geneList)    
 } else{geneDF <- MonocyteSubtype.markers_sig}
  pct_df <- melt(geneDF, id.vars = c("Gene","avg_logFC"), measure.vars = c("pct.2", "pct.1"), 
                 variable.name = "Cluster", value.name = "FractionCells") 
  pct_df$Cluster <- ifelse(pct_df$Cluster=="pct.1", "1", "0") 
  return(pct_df)
}

cellFractions_plot <- function (MonocyteSubtype.markers_sig, geneList, title="", allGenes=F){ 
  pct_df <- make_fractionDF(MonocyteSubtype.markers_sig, geneList, allGenes) 
  # Fraction Cells
  cfp <- ggplot(data=pct_df, aes(x=Gene, y=FractionCells, fill=Cluster)) + geom_col(position="dodge") + 
    labs(title = title, y="Fraction of Cells", x="Gene") +  
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
  if(allGenes==T){
    cfp <- cfp +  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 8))
    print(ggplotly(cfp))
  }else{print(cfp)}
}

logFC_plot <- function(MonocyteSubtype.markers_sig, geneList, title="", allGenes=F){
  pct_df <- make_fractionDF(MonocyteSubtype.markers_sig, geneList, allGenes) 
  # LogFC
  lfcp <- ggplot(data=pct_df, aes(x=Gene, y=avg_logFC, fill=Gene)) + geom_col(position="dodge") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1), legend.position = "none") +
     labs(title = title, y="avg_logFC", x="Gene")
   if(allGenes==T){
    lfcp <- lfcp +  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 8))
    print(ggplotly(lfcp))
  }else{print(lfcp)} 
} 

enrichment_test_and_plots <- function(DAT, DEG_df, geneList, disease=""){
  overlappingGenes <- report_overlap(clustDAT, 
                                      list1=row.names(DEG_df), 
                                      list2=geneList)
  ## All DEGs
  # cellFractions_plot(DEG_df, geneList,
  #                      title= paste("Fraction of Cells Expressing\nAll",disease,"Genes"), allGenes = T)
  # logFC_plot(DEG_df, geneList,  title= "LogFC\nAll PD Genes", allGenes = T)
  # Overlapping Genes
  if(length(overlappingGenes)>0){
    FeaturePlot(DAT, features.plot = overlappingGenes, dark.theme = T) 
    cellFractions_plot(DEG_df, overlappingGenes,
                       title= paste("Fraction of Cells Expressing\n",disease,"Genes Overlapping with DEGs") )
    logFC_plot(DEG_df, overlappingGenes, title= paste("LogFC\n",disease,"Genes Overlapping with DEGs") )
    
  
    if(length(overlappingGenes)>1){
      overlap_heatmap(clustDAT, geneList=overlappingGenes, 
                      title=paste("Overlapping Genes:\n",disease,"Genes vs. DGE Genes") )
    } 
  } else {cat("There were no overlapping genes between the DGE list and",disease,"genes.")} 
  return(overlappingGenes)
}
purpleScale <-  c("grey","purple","blueviolet","magenta")
heatScale <-  c("yellow","grey","red") 
```

### PD Genes vs. DGE Overlap

```{r PD Genes vs. DGE Overlap}  
overlappingGenes_PD <- enrichment_test_and_plots(clustDAT, MonocyteSubtype.markers_sig, PDgenes$Gene)  
```

### AD Genes vs. DGE Overlap

```{r AD Genes vs. DGE Overlap}
overlappingGenes_AD <- enrichment_test_and_plots(clustDAT, MonocyteSubtype.markers_sig, ADgenes$Gene)  
```
 

# Identify Cluster-specific Biomarkers

```{r Identify Cluster-specific Biomarkers}
identify_unique_markers <-function(DAT, clusterA, clusterB, allGenes=F){ 
  if(allGenes==F){
      clustA.markers <- FindMarkers(DAT, ident.1=clusterA, min.pct = 0.25, only.pos = F, test.use = "wilcox")
      clustB.markers <- FindMarkers(DAT, ident.1=clusterB, min.pct = 0.25, only.pos = F, test.use = "wilcox") 
  }else{
      clustA.markers <- FindMarkers(DAT, ident.1=clusterA, min.pct = 0.25, only.pos = F, test.use = "wilcox", 
                        logfc.threshold = 0, min.pct = 0, min.cells.group = 1, 
                        min.cells.gene = 1, min.diff.pct = -Inf)
      clustB.markers <- FindMarkers(DAT, ident.1=clusterB, min.pct = 0.25, only.pos = F, test.use = "wilcox", 
                        logfc.threshold = 0, min.pct = 0, min.cells.group = 1, 
                        min.cells.gene = 1, min.diff.pct = -Inf)
  } 
  clustA.uniqueMarkers <- clustA.markers[!(row.names(clustA.markers) %in% row.names(clustB.markers)),] %>%
    subset(p_val_adj<=0.05)
  clustB.uniqueMarkers <- clustB.markers[!(row.names(clustB.markers) %in% row.names(clustA.markers)),] %>%
    subset(p_val_adj<=0.05)
  
  difference <- abs( length(row.names(clustA.uniqueMarkers)) - length(row.names(clustB.uniqueMarkers) ) )
  uniqueMarkers <- data.frame(Cluster0_markers=c(row.names(clustA.uniqueMarkers), rep("",difference) ),
                              Cluster1_markers=row.names(clustB.uniqueMarkers))
  write.csv(uniqueMarkers,
            file.path(resultsPath,"unique_cluster_markers.csv"), 
            quote = F, row.names = F)
  createDT(uniqueMarkers, "Unique/Mutually Exclusive Markers of Cluster 0 and Cluster 1") 
  return(uniqueMarkers)
}



uniqueMarkers <- identify_unique_markers(DAT, clusterA = 0, clusterA = 1, allGenes=allGenes)
```
 
 
# DGE Across Clusters

## Disease

```{r DGE Across Clusters: dx, results='asis'}
# clustDAT@meta.data$dx %>% unique()
# FDR <- 0.05/dim(DEGs)[1]
# subset(DEGs, p_val<FDR)
DEGs <- runDGE(clustDAT, meta_var = "dx", group1 = "PD", group2 = "control", 
               allGenes = allGenes) 
```

## Mutation

```{r DGE Across Clusters: mut, results='asis'}
# clustDAT@meta.data$dx %>% unique()
DEGs <- runDGE(clustDAT, meta_var = "mut", group1 = "PD", group2 = "GBA", 
               allGenes = allGenes)
```


# DGE Within Clusters 

## Disease {.tabset .tabset-fade .tabset-pills}

```{r Within Clusters: dx, results='asis'}  
DGE_within_clusters(clustDAT,  meta_var = "dx", group1 = "PD", group2 = "control", clusterList = c(0,1), 
               allGenes = allGenes)
```

## Mutation {.tabset .tabset-fade .tabset-pills}

```{r Within Clusters: mut, results='asis'}  
DGE_within_clusters(DAT, meta_var = "mut", group1 = "PD", group2 = "GBA", clusterList = c(0,1), 
               allGenes = allGenes)
```

 
 
# DEG Enrichment w/ enrichR {.tabset .tabset-fade .tabset-pills}

```{r DEG Enrichment, results='asis'}
enrichr_dbs <- c("KEGG_2018", "Reactome_2016",
                 "GO_Biological_Process_2018", "GO_Molecular_Function_2018", "GO_Cellular_Component_2018", 
                 "Rare_Diseases_AutoRIF_ARCHS4_Predictions", "Human_Gene_Atlas")
# createDT(enrichR::listEnrichrDbs(), "Enrichr Databases")
 
geneList <- data.frame(Gene=row.names(MonocyteSubtype.markers), 
     Weight=scales::rescale(length(MonocyteSubtype.markers$p_val_adj):1))

results <- enrichr(genes = geneList, databases = enrichr_dbs ) 

for (db in enrichr_dbs){
  cat('\n')
  cat("##",db,"\n")  
  # res <- subset(results[[db]], Adjusted.P.value<=0.05)
  createDT_html(results[[db]], paste("Enrichr Results:",db))
  cat('\n')
}  
```





