---
title: "PD_scRNAseq: Monocle3" 
subtitle: "DGE + Top Specific Markers"
author: "<h3>Author</h3>Brian M. Schilder, Bioinformatician II"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document: 
    theme: spacelab
    highlight: zenburn 
    code_folding: show 
    toc: true 
    toc_float: true
    smooth_scroll: true
    number_sections: false 
    self_contained: true
params:
  subsetGenes: "protein_coding" #FALSE
  subsetCells: 500 #FALSE
  resolution: 0.6
  resultsPath: "./Results"
  nCores: 2
  perplexity: 30
editor_options: 
  chunk_output_type: inline
---

# Setup 

```{r setup, dpi = 600, warning=F, message=F}
start_time <- Sys.time() 

#### #### Load Objects & Functions #### #### 
######################################################
# Import functions
library(monocle3) # devtools::install_github('cole-trapnell-lab/monocle3')
paste("monocle3", packageVersion("monocle3"))  

root = "./"
source(file.path(root,"MAIN.R"))
import_parameters(params)
######################################################

#### #### PACKAGES #### #### 
###################################################### 

library(cowplot) 
library(ggplot2)
library(dplyr) 
library(data.table) 
library(readxl) 
library(reshape2)
library(ggrepel)
library(plotly)
library(GeneOverlap) # BiocManager::install("GeneOverlap") 
######################################################


# Exporting 3D plots 
knitr::knit_hooks$set(webgl = rgl::hook_webgl)
dge_limit <- F # 100
nCores <- 4#parallel::detectCores()
set.seed(2019)

# Load preprocessed data
load("./Data/monocle3_CDS.RData")
cds
```

```{r}
plt_umap <- monocle3::plot_cells(cds, show_trajectory_graph = F, group_cells_by = "cluster",
                                 alpha = .8, group_label_size = 5) 
plt_umap
ggsave("./Results/UMAP_clusters_plot.png", width=5, height=5) 

plt_umap3d <- monocle3::plot_cells_3d(cds, show_trajectory_graph = F)
print(plt_umap3d)

```


# Differential Gene Expression

Detect whether or not to limit DGE analysis to only the top N most variable genes.

```{r Differential Gene Expression}
# DGE limiter
# dge_limit <- 10
if(dge_limit){
  cds_DGE <- cds[var.genes[1:dge_limit],]
  print(paste0("Testing only the top ",dge_limit," most variable genes in DGE analysis."))
} else{
  cds_DGE <- cds
  print(paste0("Testing all ",dim(cds)[1]," genes in DGE analysis."))
} 
# DGE distribution type
expression_family <- "quasipoisson" # Recommended by Monocle3 authors for most cases.
within_clusters <- c(1,2)
DT_max <- 1000
```

## Alternative DGE Methods

```{r} 
library(Seurat)
library(future)
plan(strategy = "multicore", workers = 4) 

load("Results/3-11-2019/scRNAseq_results.RData")
Seurat::as.Seurat(mDAT)
mDAT <- Seurat::UpdateSeuratObject(mDAT)
markers <- FindMarkers(mDAT, 
                       ident.1 = "CD14+ Mono", 
                       ident.2 = "FCGR3A+ Mono", )



# degs <- read.csv("~/Downloads/PD_scRNAseq Monocle3.csv", row.names = 1)
# towfique_genes = c("MS4A4A", "MYL6", "NDUFB1","PGK1","RPL35A","RPL39",
#                    "S100A4","S100A6","SH2D3C","TAGLN")
# tp <- plot_cells(cds, genes = towfique_genes,#degs$gene_short_name[1:10], 
#                  show_trajectory_graph = F, alpha = .001, cell_size = .5)
# tp
# ggsave("./Results/curated_modules/towfique_genes.png", width = 10)
 
# MEDIANS
all_genes <- colnames(t(exprs(cds)))#row.names(exprs(cds))
all_res <- lapply(row.names(exprs(cds)), function(gene){
    print(gene)
    cds_gene <- cds[gene,]
    gene_df <- data.frame(t(exprs(cds_gene)), pData(cds))
    colnames(gene_df)[1] <- "expression"
    gene_medians <- aggregate(expression ~ ID + dx + Cluster, gene_df, median, na.rm=T) 
    res <- stats::wilcox.test(expression ~ dx, data=gene_medians)
    res_df <- data.frame(stat=res$statistic, pval=res$p.value, row.names = gene)
    return(res_df)
}) %>% data.table::rbindlist()

```




## Across Clusters

### PD vs. Control

```{r Across Clusters: PD vs. Control} 
dge.dx <- monocle3_DGE(cds_DGE = cds_DGE,
                           variable = "dx",
                           nCores = nCores, 
                           expression_family = expression_family,
                           plot_volcano = T, 
                           results_path = "./Results/across_PD.vs.Ctrl.csv") 
createDT(head(dge.dx, DT_max))
```

### PD vs. GBA

```{r Across Clusters: PD vs. GBA} 
dge.mut <- monocle3_DGE(cds_DGE = cds_DGE,
                           variable = "mut",
                           variable_subsets = c("PD","GBA"),
                           nCores = nCores, 
                           expression_family = expression_family,
                           plot_volcano = T, 
                           results_path = "./Results/across_PD.vs.GBA.csv") 
createDT(head(dge.mut, DT_max))
```

### Canonical vs. Intermediate Monocytes

- Compare all Canonical Monocytes (**Cluster 1**) to all Intermediate Monocytes (**Cluster 2**), regardless of disease status.

```{r Across Clusters: Canonical vs. Intermediate Monocytes}
dge.clust <- monocle3_DGE(cds_DGE = cds_DGE,
                          variable = "Cluster", 
                          variable_subsets = c(1,2),
                          nCores = nCores, 
                          expression_family = expression_family,
                          plot_volcano = T, 
                          results_path = "./Results/across_Clust1.vs.Clust2.csv") 
createDT(head(dge.clust, DT_max))
```


## Within Clusters

### Disease {.tabset .tabset-fade .tabset-pills}

- Compare PD vs. Controls DGE within each cluster.  
- **Cluster 1** = Canonical Monocytes, **Cluster 2** = Intermediate Monocytes.  

```{r Within Clusters: Disease, results="asis"}
for (clust in within_clusters){
  cat("\n")
  cat("#### Cluster ",clust,"\n")
  cds_clust <- cds_DGE[,pData(cds_DGE)$Cluster==clust]
  dge.dx.clust <- monocle3_DGE(cds_DGE = cds_clust,
                           variable = "dx",  
                           nCores = nCores, 
                           expression_family = expression_family,
                           plot_volcano = T, 
                           results_path = paste0("./Results/within.Clust",clust,"_PD.vs.Ctrl.csv") )
  createDT_html(head(dge.dx.clust, DT_max)) %>% print()
  cat("\n")
}
```

### PD vs. GBA {.tabset .tabset-fade .tabset-pills}

```{r Within Clusters: Mutations, results="asis"}
for (clust in within_clusters){
  cat("\n")
  cat("#### Cluster ",clust,"\n")
  cds_clust <- cds_DGE[,(pData(cds_DGE)$Cluster==clust)]
  dge.mut.clust <- monocle3_DGE(cds_DGE = cds_clust,
                           variable = "mut",  
                           variable_subsets = c("PD","GBA"),
                           nCores = nCores, 
                           expression_family = expression_family,
                           plot_volcano = T, 
                           results_path = paste0("./Results/within.Clust",clust,"_PD.vs.GBA.csv") ) 
  createDT_html(head(dge.mut.clust, DT_max)) %>% print()
  cat("\n")
}
```


# Top Specific Markers

- Find genes that characterize each cluster.  
- Specifically, only compare Clusters 1 and 2 to focuses on differences between these.  

```{r Top Specific Markers, echo=T, message=FALSE}
# Cluster-specific markers
marker_res <- top_cluster_markers(cds_DGE, 
                                  cluster_list=c(1,2), 
                                  genes_to_test_per_group=1000, 
                                  save_path= "./Results/cluster_markers.csv") 
# marker_res <- data.table::fread("./Results/cluster_markers.csv")
```

## Results Table

When trying to identify cluster-specific markers you need to consider the other columns in the DGE results table, namely:  
- ``marker_score``  
- ``fraction_expressing``  
- ``specificity``  
- ``pseudo_R2``  

```{r Top Specific Markers - Results Table}
filtered_res <- marker_res %>%
    filter(fraction_expressing >= 0.10 & marker_test_q_value <=0.05)  %>% 
  dplyr::mutate(neg_log10_q_value = -log10(marker_test_q_value)) %>%
    group_by(cell_group) %>% arrange(desc(specificity), desc(pseudo_R2))  
createDT(filtered_res)
```

## Alternative Plots

- For genes further down the properly sorted *Top Specific Markers* table may be more difficult to see in UMAP gene expression plots due to a variety of reasons (e.g. excessive cell density, low fraction of cells expressing).  
- Thus, it can be more useful to produce alternative means of visualizing these genes.  

```{r Top Specific Markers - Alternative Plots, fig.height=8, fig.width=14}  
select.genes <- c("COX5A", "COX6B1", "COX7A2", "COX7B", "NDUFA1", "MS4A4A", "PYCARD", "S100A4")
shared.metrics <- c("pseudo_R2", "neg_log10_q_value")
cluster_specific.metrics <- c("mean_expression","marker_score","fraction_expressing", "specificity")

shared <-  data.table::melt.data.table(data.table(subset(filtered_res, gene_id %in% select.genes)), 
                                                    id.vars = c("gene_id","cell_group"), 
                                                    measure.vars = shared.metrics, 
                                                    variable.name = "metric")
cluster_specific <-  data.table::melt.data.table(data.table(subset(filtered_res, gene_id %in% select.genes)), 
                                                    id.vars = c("gene_id","cell_group"), 
                                                    measure.vars = cluster.specific.metrics, 
                                                    variable.name = "metric")


g1 <- ggplot(shared) + 
      geom_col(aes(x=gene_id, y=value, fill=factor(gene_id)), show.legend = F) + 
      facet_grid(metric~.,scales = "free_y")

g2 <- ggplot(cluster_specific, aes(x=gene_id, y=value, fill=factor(cell_group))) + 
      geom_col(position = "dodge") + 
      facet_grid(metric~.,scales = "free_y") +  scale_fill_discrete(name = "Cluster")
cowplot::plot_grid(g1, g2, labels = c("Metrics Shared by Clusters", "Metrics Specific to Clusters")) 
```


# Gene Expression Plots

## Selected Genes

```{r Selected Genes}
# Selected Marker Genes
m1 <- monocle3::plot_cells(cds, 
                     genes=c("COX5A", "COX7B", "NDUFA1", "MS4A4A", "PYCARD", "S100A4"), 
                     # group_cells_by = "cluster",
                     show_trajectory_graph = F)
ggsave2("./Results/selected_DEGs.png", m1, dpi = 1000, height = 4)
print(m1) 

# COX genes
m2 <- monocle3::plot_cells(cds, 
                     genes=c("COX5A", "COX6B1", "COX7A2", "COX7B"), 
                     # group_cells_by = "cluster",
                     show_trajectory_graph = F)
ggsave2("./Results/COX-genes.png", m2, dpi = 1000,  height = 4, width = 6)
print(m2)
```

## Cluster-specific Markers

```{r Cluster-specific Markers} 
# marker_res <- data.table::fread("./Results/cluster_markers.csv")   

# Cluster1 specific
c1 <- monocle3::plot_cells(cds, 
                     genes=(subset(marker_res, cell_group==1)%>% arrange(desc(marker_score)))$gene_id[1:6]  , 
                     # group_cells_by = "cluster",
                     show_trajectory_graph = F)
# Cluster2 specific
c2 <- monocle3::plot_cells(cds, 
                     genes=(subset(marker_res, cell_group==2) %>% arrange(desc(marker_score)))$gene_id[1:6], 
                     # group_cells_by = "cluster",
                     show_trajectory_graph = F)
cp <- cowplot::plot_grid(c1, c2, nrow = 2)
print(cp)
ggsave2("./Results/cluster_markers.png", cp, dpi = 1000, scale = 1, height = 7)
```



