---
title: "Differential Gene Expression" 
author: "<h3>Author</h3>Brian M. Schilder, Bioinformatician II"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document: 
    theme: spacelab
    highlight: zenburn 
    code_folding: show 
    toc: true 
    toc_float: true
    smooth_scroll: true
    number_sections: false 
    self_contained: true
params:
  subsetGenes: "protein_coding" #FALSE
  subsetCells: 500 #FALSE
  resolution: 0.6
  resultsPath: "./Results"
  nCores: 2
  perplexity: 30
editor_options: 
  chunk_output_type: inline
---

# Setup 

```{r setup, dpi = 600, warning=F, message=F}
start_time <- Sys.time() 
# Import functions
root = "./"
source(file.path(root,"general_functions.R"))
import_parameters(params)
load(file.path(resultsPath, "scRNAseq_results.RData"))
# resultsPath <- "Results/subsetGenes-protein_coding__subsetCells-F__Resolution-0.2__perplexity-40__nCores-4" 

library(Seurat)
library(cowplot)
library(ggplot2)
library(dplyr) 
library(data.table) 
library(readxl) 
library(reshape2)
library(ggrepel)
library(plotly)

library(GeneOverlap) # BiocManager::install("GeneOverlap") 
library(enrichR) #BiocManager::install("enrichR")

library(monocle) # BiocManager::install("monocle")   
library(garnett) # devtools::install_github("cole-trapnell-lab/garnett")
 
#  
if(getwd()=="/Users/schilder/Desktop/PD_scRNAseq"){
  allGenes <- F
  test.use <- "wilcox"
}else{
  allGenes <- T
  # MAST not install on Minerva...
  test.use <- "wilcox"}
allGenes 
```


# Convert back to Seurat

```{r Convert back to Seurat} 
CDS_to_Seurat <- function(cds, export_PC=F, export_UMAP=F, export_tSNE=F){ 
  # sum(colnames(mDAT) != colnames( mDAT@reducedDimS)) 
  ## Manually save reduced dimensions
  if(export_PC==T){
    mDAT$PC1 <- mDAT@normalized_data_projection[,1]
    mDAT$PC2 <- mDAT@normalized_data_projection[,2]
    mDAT$PC3 <- mDAT@normalized_data_projection[,3] 
  }
  if(export_UMAP==T){
    mDAT$UMAP1 <- mDAT@reducedDimA[1,]
    mDAT$UMAP2 <- mDAT@reducedDimA[2,]
    mDAT$UMAP3 <- mDAT@reducedDimA[3,]
  }
  DAT <- exportCDS(mDAT, export_to = "Seurat", export_all = T)
  DAT@scale.data <- DAT@data #Data was already scaled in Monocle
  # DAT <- Seurat::AddMetaData(DAT, pData(mDAT)[c("garnett_cluster","cell_type","cluster_ext_type")])
  return(DAT)
}
DAT <- CDS_to_Seurat(mDAT, export_PC = T, export_UMAP = T)
head(DAT@meta.data) 
```


## Biomarker Expression

```{r Biomarker Expression} 
make_markerDF <- function(DAT, markerList){
  markerData <- DAT@data[row.names(DAT@data) %in% markerList,] %>% t() %>%
    as.matrix() %>%  data.table(keep.rownames = T, key = "rn")
  markerData[markerData$markerList[1]==0,] <- NA
  markerData[markerData$markerList[2]==0,] <- NA 
  colnames(markerData) <- c("Cell",paste("Gene", rep(1:length(markerList)), sep=""))
  return(markerData)
}

gene_gene_plot <- function(DAT, markerList, colorby, title="", plot=T, legend=T, filterZeros=T){ 
  
  ## Efficiently merge using data.table
  markerData <- make_markerDF(DAT, markerList)
  dt1 <- data.table(markerData, keep.rownames = "Cell", key = "Cell") %>% copy()
  dt2 <- data.table(DAT@meta.data, keep.rownames = "Cell", key = "Cell") %>% copy()
  row.names(dt2) <- dt2$Cell
  markerDT <- dt1[dt2]
  if(filterZeros==T){markerDT <- subset(markerDT, Gene1!=0 & Gene2!=0)}
  
  p <- ggplot(data = markerDT, aes(x=Gene1, y=Gene2 )) +
    stat_density_2d(aes(fill = ..level..), 
                    geom = "polygon", colour="purple", bins = 10, size=.1) +
    geom_point(aes(color=as.factor(eval(parse(text=colorby)))), shape=16, stroke=0, size=2, alpha=.5) +
    guides(colour = guide_legend(title=colorby, override.aes = list(alpha = 1))) + 
      scale_color_manual(values = pretty_colors(mDAT, var = colorby)) +
    labs(x=markerList[1], y=markerList[2], title=title) +
    geom_smooth(method="lm")
  if(legend==F){p <- p + theme(legend.position = "none")}
  if(plot==T){print(p)}
  return(p)
} 
```

### Monocyte Markers

```{r Monocyte Markers}
p <- gene_gene_plot(DAT, c("CD14", "FCGR3A"), colorby="cluster_ext_type_filt", 
               title="Monocyte Subtype Markers")
```

### HLA Genes vs. LRRK2

```{r HLA Genes vs. LRRK2} 
# Identify HLA gene names
d <- DAT@data@Dimnames[[1]] 
HLA_genes  <- d[startsWith(d, "HLA-DR")] 
# Plot
HLA_1 <- gene_gene_plot(DAT, c("LRRK2", HLA_genes[1]), colorby="cluster_ext_type_filt", 
               title="", plot = F, legend=F)
HLA_2 <- gene_gene_plot(DAT, c("LRRK2", HLA_genes[2]), colorby="cluster_ext_type_filt", 
               title="", plot = F, legend=F)
HLA_3 <- gene_gene_plot(DAT, c("LRRK2", HLA_genes[3]), colorby="cluster_ext_type_filt", 
               title="", plot = F, legend=F)
plot_grid(HLA_1, HLA_2, HLA_3)
```

# DGE: CD16+ vs. CD16-- Monocytes

Remove filters to get __all__ DEGs.
```{r DGE: CD16+ vs. CD16-- Monocytes}
clustDAT <- SubsetData(DAT, subset.name = "Cluster", accept.value = c(1,2), do.scale = F)

DEGs_monocytes <- runDGE(DAT, meta_var = "Cluster", group1 = 1, group2 = 2, 
       allGenes = allGenes, test.use = test.use)
# DEGs_monocytes <- read.csv("Results/MonocyteSubtype.markers.csv", row.names = 1)
createDT(DEGs_monocytes, caption="DEGs between cluster 1 (CD16-- monocytes) and cluster 2 (CD16+ monocytes")
write.csv(DEGs_monocytes, file.path(resultsPath, "MonocyteSubtype.markers.csv"), quote = F, row.names = T)
```

## Identify Cluster-specific Biomarkers

```{r Identify Cluster-specific Biomarkers}
identify_unique_markers <-function(DAT, clusterA, clusterB, allGenes=F, test.use="wilcox"){ 
  DAT <- SetIdent(DAT, ident.use =  DAT@meta.data$Cluster)  
  if(allGenes==F){
      clustA.markers <- FindMarkers(DAT, ident.1=clusterA, min.pct = 0.25, 
                                    only.pos = F, test.use = test.use)
      clustB.markers <- FindMarkers(DAT, ident.1=clusterB, min.pct = 0.25, 
                                    only.pos = F, test.use = test.use) 
  }else{
      clustA.markers <- FindMarkers(DAT, ident.1=clusterA, 
                                    only.pos = F, test.use = test.use,
                                    logfc.threshold = -Inf, min.pct = -Inf, 
                                    min.cells.group = -Inf,min.cells.gene = -Inf,
                                    min.diff.pct = -Inf)
      clustB.markers <- FindMarkers(DAT, ident.1=clusterB,  
                                    only.pos = F, test.use = test.use,
                                    logfc.threshold = -Inf, min.pct = -Inf,
                                    min.cells.group = -Inf,min.cells.gene = -Inf,
                                    min.diff.pct = -Inf)
  } 
  clustA.markers <- cbind(Gene=row.names(clustA.markers), clustA.markers)
  clustB.markers <- cbind(Gene=row.names(clustB.markers), clustB.markers)
  
  clustA.uniqueMarkers <- clustA.markers[!(row.names(clustA.markers) %in% row.names(clustB.markers)),] %>%
    subset(p_val_adj<=0.05) %>% mutate(Cluster=clusterA)
  clustB.uniqueMarkers <- clustB.markers[!(row.names(clustB.markers) %in% row.names(clustA.markers)),] %>%
    subset(p_val_adj<=0.05) %>% mutate(Cluster=clusterB)
  
  uniqueMarkers <- rbind(clustA.uniqueMarkers, clustB.uniqueMarkers)
  row.names(uniqueMarkers) <- uniqueMarkers$Gene
  # difference <- abs( length(row.names(clustA.uniqueMarkers)) -
  #                      length(row.names(clustB.uniqueMarkers) ) )
  # uniqueMarkers <- data.frame(ClusterA_markers=c(row.names(clustA.uniqueMarkers), rep("",difference) ),
  #                             ClusterB_markers=row.names(clustB.uniqueMarkers))
  write.csv(uniqueMarkers,
            file.path(resultsPath,"unique_cluster_markers.csv"), 
            quote = F, row.names = F)
  createDT(uniqueMarkers, "Unique/Mutually Exclusive Markers of Cluster 1 and Cluster 2") 
  return(uniqueMarkers)
}
# MUST use full dataset w/ all clusters (DAT) 
uniqueMarkers <- identify_unique_markers(DAT, clusterA = 1, clusterB = 2, 
                                         allGenes=T, test.use = test.use) 
```


# AD/PD-related Genes

    Q: Are AD/PD-related gene expression more prevalent in classical 
    or intermediate monocyte subtypes? 
    
## Import AD/PD Gene Lists

```{r Import AD/PD Gene Lists}
# Combine AD gene lists
curatedGenes <- read_excel("Data/curated_AD-PDgene_lists.xlsx")
AD_related_genes <- read_excel(file.path(root, "Data/AD-related_genes.xlsx"))
ADgenes <- rbind(AD_related_genes, data.frame(Gene=curatedGenes$AD_panel, Category="AD") ) %>%
  unique()
# Combine PD gene lists
PDgenes <- data.frame(Gene=c("LRRK2","GBA",
                             curatedGenes$gwas_Nearest_gene_Nalls_2019, 
                             curatedGenes$QTL_Nominated_genes_Nalls_2019,
                             curatedGenes$TWAS_Garrett
                             ), Category="PD") %>% unique()
# Import previously run DGE results between Canonical and Intermediate monocytes
DEGs_monocytes <- read.csv(file.path(root, resultsPath,
                                     "MonocyteSubtype.markers.csv"), 
                           row.names = 1)
DEGs_monocytes_sig <- subset(DEGs_monocytes, p_val_adj <= 0.05) 
```

## Combine Expression and Metadata

```{r Combine Expression and Metadata}
combine_exprs_and_metadata <- function(DAT, geneList, resultsPath, writeCSV=T){
  exp_dat <- DAT@data[row.names(DAT@data) %in% geneList, ] %>% 
  t() %>% as.matrix() %>% data.table(keep.rownames = "Cell", key = "Cell")
  meta_dat <- DAT@meta.data %>% data.table( keep.rownames = "Cell", key = "Cell")
  combined_DAT <- meta_dat[exp_dat] 
  if(writeCSV==T){ 
    write.csv(combined_DAT, file.path(resultsPath, "AD-PD_gene_expression.csv"),
              quote = F, row.names = F)
    } 
  return(combined_DAT)
}
# combined_DAT <-combine_exprs_and_metadata(DAT, c(ADgenes$Gene, PDgenes$Gene), resultsPath)
```



## Test Gene Overlap

### Functions

#### Gene Overlap Functions

    * __"The GeneOverlap class formulates the problem as testing whether two variables are independent, which can be represented as a contingency table, and then uses Fisher’s exact test to find the statistical significance."__  
    * Low p-value means the overlap is highly significant.  
    * [Documentation](http://bioconductor.org/packages/release/bioc/vignettes/GeneOverlap/inst/doc/GeneOverlap.pdf)
```{r Test Gene Overlap} 
report_overlap <- function(DAT, list1, list2){ 
  genomeSize <- dim(DAT@raw.data)[1]
  go.obj <- newGeneOverlap(listA = list1, listB = list2, genome.size = genomeSize )
  go.obj <- testGeneOverlap(go.obj)
  print(go.obj) 
  
  overlappingGenes <- getIntersection(go.obj)
  percent_of_targetGenes <- length(overlappingGenes) / length(list2)*100
  percent_of_DEGs <- length(overlappingGenes) / length(list1)*100
  targetGenes_DEGs <- list2[overlappingGenes %in% list2]
  
  cat("\n",round(percent_of_targetGenes, 2),"% of the provided genes (",length(overlappingGenes),"/",length(list2),") are differentially expressed between Canonical vs. Intermediate monocyte subtypes.")
   
  cat("\n",round(percent_of_DEGs, 2),"% of the DEGs between Canonical vs. Intermediate  monocyte subtypes (",
      length(overlappingGenes),"/",length(list1),") are in the provided gene list.")
  
  cat("\n-------------------------------------------------------",
      "\n\n++++++++++++ Enrichment p-value =",go.obj@pval,"++++++++++++\n\n")
  return(overlappingGenes) 
} 
```

#### Overlap Plot Functions

```{r Overlap Plot Functions}
###############################################
# PLOT FUNCTIONS
###############################################

# HEATMAP
overlap_heatmap <- function(clustDAT, geneList, 
                            title="Overlapping Genes:\nGene List vs. DGE Genes", 
                            interactive=F){ 
  knitr::opts_chunk$set(fig.width=14, fig.height=14) 
  markerDF  <- get_markerDT(clustDAT, markerList = geneList, rawData = F)
  markerMatrix <- reshape2::acast(markerDF, Gene~Cluster, value.var="Expression",
                                  fun.aggregate = mean, drop = F, fill = 0)
  # Heatmap.2 
  my_palette <- colorRampPalette(c("purple", "black", "cyan"))(n = 1000)
  hmap <- gplots::heatmap.2(markerMatrix, xlab = "Cluster", dendrogram = "row",
                    col = my_palette, tracecol = "gray", srtCol = 0, offsetCol=1.5, vline=T, 
                    trace='none', key.title=NA, key.ylab = "Expression", colsep=T, sepwidth = 0.01, 
                    main = title) 
   knitr::opts_chunk$set(fig.width=7, fig.height=5) 
}

# CEll FRACTIONS PLOT
make_fractionDF <- function(sig_DEGs_df, geneList, allGenes=F){
  sig_DEGs_df$Gene <- row.names(sig_DEGs_df)
 if(allGenes==F){
   geneDF <- subset(sig_DEGs_df, Gene %in% geneList)    
 } else{geneDF <- sig_DEGs_df}
  pct_df <- melt(geneDF, id.vars = c("Gene","avg_logFC"), measure.vars = c("pct.2", "pct.1"), 
                 variable.name = "Cluster", value.name = "FractionCells") 
  pct_df$Cluster <- ifelse(pct_df$Cluster=="pct.1", "1", "2") 
  return(pct_df)
}

cellFractions_plot <- function (sig_DEGs_df, 
                                geneList, title="", allGenes=F){ 
  knitr::opts_chunk$set(fig.width=14, fig.height=14)  
  pct_df <- make_fractionDF(sig_DEGs_df, geneList, allGenes) 
  # Fraction Cells
  cfp <- ggplot(data=pct_df, aes(x=Gene, y=FractionCells, fill=Cluster)) + geom_col(position="dodge") + 
    labs(title = title, y="Fraction of Cells", x="Gene") +  
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
  if(allGenes==T){
    cfp <- cfp +  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 8))
    print(ggplotly(cfp))
  }else{print(cfp)}
  knitr::opts_chunk$set(fig.width=7, fig.height=5)  
}

# LOGFC PLOT
logFC_plot <- function(sig_DEGs_df, geneList, title="", allGenes=F){
  knitr::opts_chunk$set(fig.width=14, fig.height=14)  
  pct_df <- make_fractionDF(sig_DEGs_df, geneList, allGenes)  
  lfcp <- ggplot(data=pct_df, aes(x=Gene, y=avg_logFC, fill=Gene)) + geom_col(position="dodge") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1), legend.position = "none") +
     labs(title = title, y="avg_logFC", x="Gene")
   if(allGenes==T){
    lfcp <- lfcp +  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 8))
    print(ggplotly(lfcp))
   }else{print(lfcp)} 
  knitr::opts_chunk$set(fig.width=7, fig.height=5)  
} 

enrichment_test_and_plots <- function(DAT, sig_DEGs_df, geneList, disease=""){ 
  cat("\nCalculate enrichment:\n")
  overlappingGenes <- report_overlap(clustDAT, 
                                      list1=row.names(sig_DEGs_df), 
                                      list2=geneList)
  if(length(overlappingGenes)>0){
    cat("\nPlot cell fractions\n")
    cellFractions_plot(sig_DEGs_df, overlappingGenes,
                       title= paste("Fraction of Cells Expressing\n",
                                    disease,"Genes Overlapping with DEGs") )
    cat("\nPlot logFC")
    logFC_plot(sig_DEGs_df, overlappingGenes, 
               title= paste("LogFC\n",disease,"Genes Overlapping with DEGs") )
    cat("\nPlot expression of all overlapping PD genes\n")
    knitr::opts_chunk$set(fig.width=10, fig.height=20)  
    plot_cell_clusters(mDAT, markers = overlappingGenes, cell_size = 0.7) 
    
    if(length(overlappingGenes)>1){ 
      cat("\nPlot heatmap\n")
      overlap_heatmap(clustDAT, geneList=overlappingGenes, 
                      title=paste("Overlapping Genes:\n",
                                  disease,"Genes vs. DGE Genes") )
    } 
  } else {cat("There were no overlapping genes between the DGE list and",disease,"genes.")} 
  return(overlappingGenes)
}
# purpleScale <-  c("grey","purple","blueviolet","magenta")
# heatScale <-  c("yellow","grey","red") 
```

### PD Genes  

### PD Genes vs. Overlapping DGEs

```{r PD Genes vs. Overlapping DGEs}  
overlappingGenes_PD <- enrichment_test_and_plots(clustDAT, DEGs_monocytes_sig,
                                                 PDgenes$Gene) 
overlap_heatmap(clustDAT, geneList = PDgenes$Gene, title = "All PD-related Genes")
```

### PD Genes vs. Unique Cluster Markers

```{r PD Genes vs. Overlapping Unique Cluster Markers}  
Clust1Markers_PDgenes <- enrichment_test_and_plots(clustDAT, subset(uniqueMarkers, Cluster==1), PDgenes$Gene)

Clust2Markers_PDgenes <- enrichment_test_and_plots(clustDAT, subset(uniqueMarkers, Cluster==2), PDgenes$Gene)
```


### AD Genes vs. DGE Overlap

```{r AD Genes vs. DGE Overlap}
overlappingGenes_AD <- enrichment_test_and_plots(clustDAT, DEGs_monocytes_sig, ADgenes$Gene)
overlap_heatmap(clustDAT, geneList = ADgenes$Gene, title = "All AD-related Genes")
```

### AD Genes vs. Unique Cluster Markers

```{r AD Genes vs. Overlapping Unique Cluster Markers}  
Clust1Markers_ADgenes <- enrichment_test_and_plots(clustDAT, subset(uniqueMarkers, Cluster==1), ADgenes$Gene)

Clust2Markers_ADgenes <- enrichment_test_and_plots(clustDAT, subset(uniqueMarkers, Cluster==2), ADgenes$Gene)
```




# DGE Across Clusters {.tabset .tabset-fade .tabset-pills}

## Disease

```{r DGE Across Clusters: dx, results='asis'}
# clustDAT@meta.data$dx %>% unique()
# FDR <- 0.05/dim(DEGs)[1]
# subset(DEGs, p_val<FDR)
DEGs <- runDGE(clustDAT, meta_var = "dx", group1 = "PD", group2 = "control", 
               allGenes = F, test.use = test.use) 
```

## Mutation

```{r DGE Across Clusters: mut, results='asis'}
# clustDAT@meta.data$dx %>% unique()
DEGs <- runDGE(clustDAT, meta_var = "mut", group1 = "PD", group2 = "GBA", 
               allGenes = F, test.use = test.use)
```


# DGE Within Clusters 

## Disease {.tabset .tabset-fade .tabset-pills}

```{r Within Clusters: dx, results='asis'}  
DGE_within_clusters(clustDAT,  meta_var = "dx", group1 = "PD", group2 = "control", 
                    clusterList = c(1,2), allGenes = F, test.use = test.use)
```

## Mutation {.tabset .tabset-fade .tabset-pills}

```{r Within Clusters: mut, results='asis'}  
DGE_within_clusters(DAT, meta_var = "mut", group1 = "PD", group2 = "GBA", 
                    clusterList = c(1,2), allGenes = F, test.use = test.use)
```

 
 
# DEG Enrichment w/ enrichR {.tabset .tabset-fade .tabset-pills}

```{r DEG Enrichment, results='asis'}
enrichr_dbs <- c("KEGG_2018", "Reactome_2016",
                 "GO_Biological_Process_2018", "GO_Molecular_Function_2018", "GO_Cellular_Component_2018", 
                 "Rare_Diseases_AutoRIF_ARCHS4_Predictions", "Human_Gene_Atlas")
# createDT(enrichR::listEnrichrDbs(), "Enrichr Databases")
 
geneList <- data.frame(Gene=row.names(DEGs_monocytes), 
     Weight=scales::rescale(length(DEGs_monocytes$p_val_adj):1))

results <- enrichr(genes = geneList, databases = enrichr_dbs ) 

for (db in enrichr_dbs){
  cat('\n')
  cat("##",db,"\n")  
  # res <- subset(results[[db]], Adjusted.P.value<=0.05)
  createDT_html(results[[db]], paste("Enrichr Results:",db))
  cat('\n')
}  
```

# Save Results

```{r Save Results}  
save.image(file.path(resultsPath, "scRNAseq_results.RData"))

end_time <- Sys.time()
end_time - start_time
cat("\n\n")
sessionInfo()
```

