<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2019-08-13" />

<title>Enrichment</title>

<script src="scRNAseq_Monocle3_Enrich_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="scRNAseq_Monocle3_Enrich_files/bootstrap-3.3.5/css/spacelab.min.css" rel="stylesheet" />
<script src="scRNAseq_Monocle3_Enrich_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="scRNAseq_Monocle3_Enrich_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="scRNAseq_Monocle3_Enrich_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="scRNAseq_Monocle3_Enrich_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="scRNAseq_Monocle3_Enrich_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="scRNAseq_Monocle3_Enrich_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="scRNAseq_Monocle3_Enrich_files/navigation-1.1/tabsets.js"></script>
<script src="scRNAseq_Monocle3_Enrich_files/navigation-1.1/codefolding.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  { color: #cccccc; background-color: #303030; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ffcfaf; } /* Alert */
code span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #dca3a3; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #f0dfaf; } /* ControlFlow */
code span.ch { color: #dca3a3; } /* Char */
code span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code span.co { color: #7f9f7f; } /* Comment */
code span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code span.do { color: #7f9f7f; } /* Documentation */
code span.dt { color: #dfdfbf; } /* DataType */
code span.dv { color: #dcdccc; } /* DecVal */
code span.er { color: #c3bf9f; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #c0bed1; } /* Float */
code span.fu { color: #efef8f; } /* Function */
code span.im { } /* Import */
code span.in { color: #7f9f7f; font-weight: bold; } /* Information */
code span.kw { color: #f0dfaf; } /* Keyword */
code span.op { color: #f0efd0; } /* Operator */
code span.ot { color: #efef8f; } /* Other */
code span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code span.sc { color: #dca3a3; } /* SpecialChar */
code span.ss { color: #cc9393; } /* SpecialString */
code span.st { color: #cc9393; } /* String */
code span.va { } /* Variable */
code span.vs { color: #cc9393; } /* VerbatimString */
code span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Enrichment</h1>
<h4 class="author"><h3>
Author
</h3>
Brian M. Schilder, Bioinformatician II</h4>
<h4 class="date">2019-08-13</h4>

</div>


<p>Enrichment tests on differentially expressed genes between Canonical monocytes (cluster 1) and Intermediate monocytes (cluster 1).</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># Import functions</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">root =<span class="st"> &quot;./&quot;</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw">source</span>(<span class="kw">file.path</span>(root,<span class="st">&quot;MAIN.R&quot;</span>))</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"></a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="kw">library</span>(enrichR) <span class="co"># BiocManager::install(&quot;enrichR&quot;) # devtools::install_github(&quot;wjawaid/enrichR&quot;)</span></a></code></pre></div>
<pre><code>## Welcome to enrichR
## Checking connection ... Connection is Live!</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">library</span>(fgsea)</a></code></pre></div>
<pre><code>## Loading required package: Rcpp</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">library</span>(reactome.db) <span class="co"># BiocManager::install(&quot;reactome.db&quot;)</span></a></code></pre></div>
<pre><code>## Loading required package: AnnotationDbi
## Loading required package: stats4
## Loading required package: BiocGenerics
## Loading required package: parallel
## 
## Attaching package: &#39;BiocGenerics&#39;
## 
## The following objects are masked from &#39;package:parallel&#39;:
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB
## 
## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs
## 
## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, basename, cbind,
##     colnames, dirname, do.call, duplicated, eval, evalq, Filter,
##     Find, get, grep, grepl, intersect, is.unsorted, lapply, Map,
##     mapply, match, mget, order, paste, pmax, pmax.int, pmin,
##     pmin.int, Position, rank, rbind, Reduce, rownames, sapply,
##     setdiff, sort, table, tapply, union, unique, unsplit, which,
##     which.max, which.min
## 
## Loading required package: Biobase
## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     &#39;browseVignettes()&#39;. To cite Bioconductor, see
##     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.
## 
## Loading required package: IRanges
## Loading required package: S4Vectors
## 
## Attaching package: &#39;S4Vectors&#39;
## 
## The following object is masked from &#39;package:base&#39;:
## 
##     expand.grid</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw">library</span>(dplyr) </a></code></pre></div>
<pre><code>## 
## Attaching package: &#39;dplyr&#39;
## 
## The following object is masked from &#39;package:AnnotationDbi&#39;:
## 
##     select
## 
## The following objects are masked from &#39;package:IRanges&#39;:
## 
##     collapse, desc, intersect, setdiff, slice, union
## 
## The following objects are masked from &#39;package:S4Vectors&#39;:
## 
##     first, intersect, rename, setdiff, setequal, union
## 
## The following object is masked from &#39;package:Biobase&#39;:
## 
##     combine
## 
## The following objects are masked from &#39;package:BiocGenerics&#39;:
## 
##     combine, intersect, setdiff, union
## 
## The following objects are masked from &#39;package:stats&#39;:
## 
##     filter, lag
## 
## The following objects are masked from &#39;package:base&#39;:
## 
##     intersect, setdiff, setequal, union</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw">library</span>(ggplot2)</a>
<a class="sourceLine" id="cb9-2" data-line-number="2"></a>
<a class="sourceLine" id="cb9-3" data-line-number="3"><span class="kw">load</span>(<span class="st">&quot;./Data/monocle3_CDS.RData&quot;</span>)</a>
<a class="sourceLine" id="cb9-4" data-line-number="4">cds</a></code></pre></div>
<pre><code>## Loading required package: monocle3
## Loading required package: SingleCellExperiment
## Loading required package: SummarizedExperiment
## Loading required package: GenomicRanges
## Loading required package: GenomeInfoDb
## Loading required package: DelayedArray
## Loading required package: matrixStats
## 
## Attaching package: &#39;matrixStats&#39;
## 
## The following object is masked from &#39;package:dplyr&#39;:
## 
##     count
## 
## The following objects are masked from &#39;package:Biobase&#39;:
## 
##     anyMissing, rowMedians
## 
## Loading required package: BiocParallel
## 
## Attaching package: &#39;DelayedArray&#39;
## 
## The following objects are masked from &#39;package:matrixStats&#39;:
## 
##     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges
## 
## The following objects are masked from &#39;package:base&#39;:
## 
##     aperm, apply, rowsum
## 
## 
## Attaching package: &#39;monocle3&#39;
## 
## The following objects are masked from &#39;package:Biobase&#39;:
## 
##     exprs, fData, fData&lt;-, pData, pData&lt;-</code></pre>
<pre><code>## class: cell_data_set 
## dim: 14827 19144 
## metadata(1): cds_version
## assays(1): counts
## rownames(14827): A1BG A2M ... ZZEF1 ZZZ3
## rowData names(1): gene_short_name
## colnames(19144): GTCATTTTCCGCATCT ATAAGAGAGGAGTTGC ...
##   CAGTAACCCGCTCGCA ACACCGGTCCGGCCCA
## colData names(23): nGene nUMI ... cell_type cluster_ext_type
## reducedDimNames(2): PCA UMAP
## spikeNames(0):</code></pre>
<div id="import-gene-lists" class="section level1">
<h1>Import Gene Lists</h1>
<div id="degs" class="section level2">
<h2>DEGs</h2>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="co"># DEGs</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2">dge.dx &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;Results/across_PD.vs.Ctrl.csv&quot;</span>)  </a>
<a class="sourceLine" id="cb12-3" data-line-number="3">dge.clust &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;Results/across_Clust1.vs.Clust2.csv&quot;</span>)  </a></code></pre></div>
</div>
</div>
<div id="ad-and-pd-associated-genes" class="section level1">
<h1>AD and PD-associated Genes</h1>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">curatedGenes &lt;-<span class="st"> </span>readxl<span class="op">::</span><span class="kw">read_excel</span>(<span class="st">&quot;Data/curated_AD-PDgene_lists.xlsx&quot;</span>)</a>
<a class="sourceLine" id="cb13-2" data-line-number="2">AD_related_genes &lt;-<span class="st"> </span>readxl<span class="op">::</span><span class="kw">read_excel</span>(<span class="kw">file.path</span>(root, <span class="st">&quot;Data/AD-related_genes.xlsx&quot;</span>))</a>
<a class="sourceLine" id="cb13-3" data-line-number="3">ADgenes &lt;-<span class="st"> </span><span class="kw">rbind</span>(AD_related_genes, <span class="kw">data.frame</span>(<span class="dt">Gene=</span>curatedGenes<span class="op">$</span>AD_panel, <span class="dt">Category=</span><span class="st">&quot;AD&quot;</span>) ) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">unique</span>()</a>
<a class="sourceLine" id="cb13-4" data-line-number="4">ADgenes &lt;-<span class="st"> </span>ADgenes[<span class="op">!</span><span class="kw">is.na</span>(ADgenes<span class="op">$</span>Gene),]</a>
<a class="sourceLine" id="cb13-5" data-line-number="5"><span class="co"># Combine PD gene lists</span></a>
<a class="sourceLine" id="cb13-6" data-line-number="6">PDgenes &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">Gene=</span><span class="kw">c</span>(<span class="st">&quot;LRRK2&quot;</span>,<span class="st">&quot;GBA&quot;</span>,</a>
<a class="sourceLine" id="cb13-7" data-line-number="7">                             curatedGenes<span class="op">$</span>gwas_Nearest_gene_Nalls_<span class="dv">2019</span>, </a>
<a class="sourceLine" id="cb13-8" data-line-number="8">                             curatedGenes<span class="op">$</span>QTL_Nominated_genes_Nalls_<span class="dv">2019</span>,</a>
<a class="sourceLine" id="cb13-9" data-line-number="9">                             curatedGenes<span class="op">$</span>TWAS_Garrett</a>
<a class="sourceLine" id="cb13-10" data-line-number="10">                             ), <span class="dt">Category=</span><span class="st">&quot;PD&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">unique</span>()</a>
<a class="sourceLine" id="cb13-11" data-line-number="11">PDgenes &lt;-<span class="st"> </span>PDgenes[<span class="op">!</span><span class="kw">is.na</span>(PDgenes<span class="op">$</span>Gene),]</a></code></pre></div>
</div>
<div id="curated-gene-lists" class="section level1">
<h1>Curated Gene Lists</h1>
<div id="alzheimers-disease" class="section level2">
<h2>Alzheimer’s Disease</h2>
<div id="disease-dges" class="section level3">
<h3>Disease DGEs</h3>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">AD_overlap &lt;-<span class="st"> </span><span class="kw">report_overlap</span>(<span class="dt">genomeSize=</span><span class="kw">dim</span>(cds)[<span class="dv">1</span>], </a>
<a class="sourceLine" id="cb14-2" data-line-number="2">                            <span class="dt">list1=</span><span class="kw">subset</span>(dge.dx, q_value<span class="op">&lt;=</span><span class="fl">0.05</span>)<span class="op">$</span>gene_short_name, </a>
<a class="sourceLine" id="cb14-3" data-line-number="3">                            <span class="dt">list2=</span>ADgenes<span class="op">$</span>Gene)</a></code></pre></div>
<pre><code>## GeneOverlap object:
## listA size=231
## listB size=76
## Intersection size=9
## Overlapping p-value=2.7e-06
## Jaccard Index=0.0
## 
##  11.69 % of the provided genes ( 9 / 77 ) are DEGs.
##  3.9 % of the DEGs ( 9 / 231 ) are in the provided gene list.
## ------------------------------------------------------- 
## 
## ++++++++++++ Enrichment p-value = 2.662927e-06 ++++++++++++</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="kw">overlap_expression_plot</span>(dge.dx, </a>
<a class="sourceLine" id="cb16-2" data-line-number="2">                        <span class="dt">geneList =</span> AD_overlap, </a>
<a class="sourceLine" id="cb16-3" data-line-number="3">                        <span class="dt">title=</span><span class="st">&quot;Overlapping AD-associated Genes / dx DGEs&quot;</span>)</a></code></pre></div>
<p><img src="scRNAseq_Monocle3_Enrich_files/figure-html/Overlap%20-%20AD%20vs.%20dx%20DGEs-1.png" width="672" /></p>
</div>
<div id="cluster-dges" class="section level3">
<h3>Cluster DGEs</h3>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">AD_overlap &lt;-<span class="st"> </span><span class="kw">report_overlap</span>(<span class="dt">genomeSize=</span><span class="kw">dim</span>(cds)[<span class="dv">1</span>], </a>
<a class="sourceLine" id="cb17-2" data-line-number="2">                            <span class="dt">list1=</span><span class="kw">subset</span>(dge.clust, q_value<span class="op">&lt;=</span><span class="fl">0.05</span>)<span class="op">$</span>gene_short_name, </a>
<a class="sourceLine" id="cb17-3" data-line-number="3">                            <span class="dt">list2=</span>ADgenes<span class="op">$</span>Gene)</a></code></pre></div>
<pre><code>## GeneOverlap object:
## listA size=927
## listB size=76
## Intersection size=26
## Overlapping p-value=2.7e-13
## Jaccard Index=0.0
## 
##  33.77 % of the provided genes ( 26 / 77 ) are DEGs.
##  2.8 % of the DEGs ( 26 / 927 ) are in the provided gene list.
## ------------------------------------------------------- 
## 
## ++++++++++++ Enrichment p-value = 2.681757e-13 ++++++++++++</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="kw">overlap_expression_plot</span>(dge.clust, </a>
<a class="sourceLine" id="cb19-2" data-line-number="2">                        <span class="dt">geneList =</span> AD_overlap, </a>
<a class="sourceLine" id="cb19-3" data-line-number="3">                        <span class="dt">title=</span><span class="st">&quot;Overlapping AD-associated Genes / Cluster DGEs&quot;</span>)</a></code></pre></div>
<pre><code>## Warning: Removed 2 rows containing missing values (position_stack).</code></pre>
<p><img src="scRNAseq_Monocle3_Enrich_files/figure-html/Overlap%20-%20AD%20vs.%20cluster%20DGEs-1.png" width="672" /></p>
</div>
</div>
<div id="parkinsons-disease" class="section level2">
<h2>Parkinson’s Disease</h2>
<div id="disease-dges-1" class="section level3">
<h3>Disease DGEs</h3>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">PD_overlap &lt;-<span class="st"> </span><span class="kw">report_overlap</span>(<span class="dt">genomeSize=</span><span class="kw">dim</span>(cds)[<span class="dv">1</span>], </a>
<a class="sourceLine" id="cb21-2" data-line-number="2">                            <span class="dt">list1=</span><span class="kw">subset</span>(dge.dx, q_value<span class="op">&lt;=</span><span class="fl">0.05</span>)<span class="op">$</span>gene_short_name, </a>
<a class="sourceLine" id="cb21-3" data-line-number="3">                            <span class="dt">list2=</span>PDgenes<span class="op">$</span>Gene)</a></code></pre></div>
<pre><code>## GeneOverlap object:
## listA size=231
## listB size=134
## Intersection size=1
## Overlapping p-value=0.88
## Jaccard Index=0.0
## 
##  0.75 % of the provided genes ( 1 / 134 ) are DEGs.
##  0.43 % of the DEGs ( 1 / 231 ) are in the provided gene list.
## ------------------------------------------------------- 
## 
## ++++++++++++ Enrichment p-value = 0.8792075 ++++++++++++</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="kw">overlap_expression_plot</span>(dge.dx, </a>
<a class="sourceLine" id="cb23-2" data-line-number="2">                        <span class="dt">geneList =</span> PD_overlap, </a>
<a class="sourceLine" id="cb23-3" data-line-number="3">                        <span class="dt">title=</span><span class="st">&quot;Overlapping PD-associated Genes / dx DGEs&quot;</span>)</a></code></pre></div>
<p><img src="scRNAseq_Monocle3_Enrich_files/figure-html/Overlap%20-%20PD%20vs.%20dx%20DGEs-1.png" width="672" /></p>
</div>
<div id="cluster-dges-1" class="section level3">
<h3>Cluster DGEs</h3>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1">PD_overlap &lt;-<span class="st"> </span><span class="kw">report_overlap</span>(<span class="dt">genomeSize=</span><span class="kw">dim</span>(cds)[<span class="dv">1</span>], </a>
<a class="sourceLine" id="cb24-2" data-line-number="2">                            <span class="dt">list1=</span><span class="kw">subset</span>(dge.clust, q_value<span class="op">&lt;=</span><span class="fl">0.05</span>)<span class="op">$</span>gene_short_name, </a>
<a class="sourceLine" id="cb24-3" data-line-number="3">                            <span class="dt">list2=</span>PDgenes<span class="op">$</span>Gene)</a></code></pre></div>
<pre><code>## GeneOverlap object:
## listA size=927
## listB size=134
## Intersection size=6
## Overlapping p-value=0.85
## Jaccard Index=0.0
## 
##  4.48 % of the provided genes ( 6 / 134 ) are DEGs.
##  0.65 % of the DEGs ( 6 / 927 ) are in the provided gene list.
## ------------------------------------------------------- 
## 
## ++++++++++++ Enrichment p-value = 0.8507007 ++++++++++++</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="kw">overlap_expression_plot</span>(dge.clust, </a>
<a class="sourceLine" id="cb26-2" data-line-number="2">                        <span class="dt">geneList =</span> PD_overlap, </a>
<a class="sourceLine" id="cb26-3" data-line-number="3">                        <span class="dt">title=</span><span class="st">&quot;Overlapping PD-associated Genes / Cluster DGEs&quot;</span>)</a></code></pre></div>
<p><img src="scRNAseq_Monocle3_Enrich_files/figure-html/Overlap%20-%20PD%20vs.%20cluster%20DGEs-1.png" width="672" /></p>
</div>
</div>
</div>
<div id="enrichr" class="section level1 tabset tabset-fade tabset-pills">
<h1>enrichR</h1>
<div id="cluster-1-vs.cluster-2" class="section level2">
<h2>Cluster 1 vs. Cluster 2</h2>
<ul>
<li>Showing plenty of enrichment on the Enrichr website, just not the enrichR R package.<br />
</li>
<li>Providing the <a href="https://amp.pharm.mssm.edu/Enrichr/enrich?dataset=d634d1a5b290e873498f65676c894b33">link to the online results here</a> for now.</li>
</ul>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="co"># enrichr.clusts &lt;- run_enrichr(DEG_table,</span></a>
<a class="sourceLine" id="cb27-2" data-line-number="2"><span class="co">#                               results_path = &quot;./Results/topEnrichR_hits.csv&quot;,   </span></a>
<a class="sourceLine" id="cb27-3" data-line-number="3"><span class="co">#                               top_genes=100,</span></a>
<a class="sourceLine" id="cb27-4" data-line-number="4"><span class="co">#                               include_weights=T</span></a>
<a class="sourceLine" id="cb27-5" data-line-number="5"><span class="co">#                               )</span></a></code></pre></div>
</div>
</div>
<div id="fgsea" class="section level1">
<h1>fGSEA</h1>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("show" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
